{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Establishing Consensus Host Taxonomy for vOTU Set  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import re,glob,os\n",
    "import pandas as pd\n",
    "import subprocess as sp\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO, SeqUtils\n",
    "import seaborn as sns\n",
    "from functools import reduce\n",
    "from Bio import Entrez\n",
    "Entrez.email = 'alexa_nicolas@berkeley.edu'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sources of host taxonomy:\n",
    "\n",
    "1. CRISPR matches\n",
    "* spacers in MAGs matched to vOTUs\n",
    "  (MAG taxonomy from GTDB)\n",
    "\n",
    "* spacers in unbinned metagenome sequences matched to vOTUs\n",
    "  (taxonomy using Kaiju)\n",
    "  \n",
    "* spacers in vOTUs matched to MAGs\n",
    "  (MAG taxonomy from GTDB)\n",
    "\n",
    "* spacer database (Shmakov et al., 2017 doi: 10.1128/mBio.01397-17) aligned to vOTUs\n",
    "  (spacer taxonomy from NCBI)\n",
    "  \n",
    "2. vOTU taxonomy predictions (using Kaiju)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Workflow:\n",
    "\n",
    "#### Establishing Kaiju-based taxonomy\n",
    "  1. unbinned mg contigs with spacers\n",
    "  2. vOTU taxonomy via kaiju\n",
    "#### MAG-based taxonomy\n",
    "  vOTU spacers matched to MAGs\n",
    "#### NCBI-based taxonomy\n",
    "  1. Previously published Spacer DB\n",
    "  2. MAG spacers matched to vOTUs\n",
    " \n",
    "### Host Linkage\n",
    "- matching unbinned spacers to vOTUs\n",
    "\n",
    "#### Cleaning up all host matches and merging into single df\n",
    "\n",
    "#### Establishing host consensus taxonomy\n",
    "\n",
    "#### Creating working definition of taxonomies to use in analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Unbinned metagenome contig taxonomy via Kaiju"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "for contigs not in MAGs containing CRISPR arrays, establish contig-level taxonomy using Kaiju"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in and format kaiju output \n",
    "kaiju_unbinned_spacers = pd.read_csv(\n",
    "'/groups/banfield/projects/environmental/4thwedge/2020/assembly.d/minced_disrat0.33/contigs_w_spacers_kaiju_out_names.tsv',\n",
    "    sep='\\t',\n",
    "    header=None,\n",
    "    names=['classified','kaiju_scaffold','ncbi_taxid','kaiju_taxonomy'])\n",
    "kaiju_unbinned_spacers = kaiju_unbinned_spacers[kaiju_unbinned_spacers['classified']=='C']\n",
    "kaiju_unbinned_spacers['kaiju_scaffold'] = kaiju_unbinned_spacers['kaiju_scaffold'].str.replace('_scaffold_min_2500_wspacers','')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# vOTU taxonomy via Kaiju"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "kaiju_vOTUs = pd.read_csv(\n",
    "'/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/viral_id/drep_viral_set/kaiju_names_out.tsv',\n",
    "    sep='\\t',\n",
    "    header=None,\n",
    "    names=['classified','kaiju_scaffold','ncbi_taxid','kaiju_taxonomy'])\n",
    "kaiju_vOTUs = kaiju_vOTUs[kaiju_vOTUs['classified']=='C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>classified</th>\n",
       "      <th>kaiju_scaffold</th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>kaiju_taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>C</td>\n",
       "      <td>4Wh2oV_72h_scaffold_118478</td>\n",
       "      <td>2560607</td>\n",
       "      <td>Mycobacterium virus TA17a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1088</th>\n",
       "      <td>C</td>\n",
       "      <td>4Wh2oV_3h_scaffold_781618</td>\n",
       "      <td>2852170</td>\n",
       "      <td>unclassified Avanivirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1163</th>\n",
       "      <td>C</td>\n",
       "      <td>4Wh2oV_0h_scaffold_2031272</td>\n",
       "      <td>2562660</td>\n",
       "      <td>unclassified Cheoctovirus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     classified              kaiju_scaffold  ncbi_taxid  \\\n",
       "984           C  4Wh2oV_72h_scaffold_118478     2560607   \n",
       "1088          C   4Wh2oV_3h_scaffold_781618     2852170   \n",
       "1163          C  4Wh2oV_0h_scaffold_2031272     2562660   \n",
       "\n",
       "                 kaiju_taxonomy  \n",
       "984   Mycobacterium virus TA17a  \n",
       "1088    unclassified Avanivirus  \n",
       "1163  unclassified Cheoctovirus  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kaiju_vOTUs[kaiju_vOTUs['kaiju_taxonomy'].str.contains('virus')].head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For vOTUs with viral taxonomies establish host using Virus-Host DB (https://www.genome.jp/ftp/db/virushostdb/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "virushostdb = pd.read_csv('/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/kaiju_out/virushostdb.tsv',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>virus tax id</th>\n",
       "      <th>virus name</th>\n",
       "      <th>virus lineage</th>\n",
       "      <th>refseq id</th>\n",
       "      <th>KEGG GENOME</th>\n",
       "      <th>KEGG DISEASE</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>host tax id</th>\n",
       "      <th>host name</th>\n",
       "      <th>host lineage</th>\n",
       "      <th>pmid</th>\n",
       "      <th>evidence</th>\n",
       "      <th>sample type</th>\n",
       "      <th>source organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10723</th>\n",
       "      <td>2560607</td>\n",
       "      <td>Mycobacterium virus TA17a</td>\n",
       "      <td>Viruses; Duplodnaviria; Heunggongvirae; Urovir...</td>\n",
       "      <td>NC_043767</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>246196.0</td>\n",
       "      <td>Mycolicibacterium smegmatis MC2 155</td>\n",
       "      <td>Bacteria; Terrabacteria group; Actinobacteria;...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCBI Virus, RefSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       virus tax id                 virus name  \\\n",
       "10723       2560607  Mycobacterium virus TA17a   \n",
       "\n",
       "                                           virus lineage  refseq id  \\\n",
       "10723  Viruses; Duplodnaviria; Heunggongvirae; Urovir...  NC_043767   \n",
       "\n",
       "      KEGG GENOME KEGG DISEASE DISEASE  host tax id  \\\n",
       "10723         NaN          NaN     NaN     246196.0   \n",
       "\n",
       "                                 host name  \\\n",
       "10723  Mycolicibacterium smegmatis MC2 155   \n",
       "\n",
       "                                            host lineage pmid  \\\n",
       "10723  Bacteria; Terrabacteria group; Actinobacteria;...  NaN   \n",
       "\n",
       "                 evidence sample type  source organism  \n",
       "10723  NCBI Virus, RefSeq         NaN              NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "virushostdb[virushostdb['virus name']==\"Mycobacterium virus TA17a\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### using ncbi_taxid to get full taxonomy for kaiju contigs predicted\n",
    "kaiju output does not provide full phylogenetic information – just highest resolution level taxonomy. Use Entrez to look up full NCBI taxonomy for kaiju output using provided ncbi taxids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_taxid = set(kaiju_vOTUs['ncbi_taxid'].unique().tolist() + kaiju_unbinned_spacers['ncbi_taxid'].unique().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ncbi_taxa_dict = {}\n",
    "with open(\"/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/kaiju_out/ncbi_taxid_kaiju_contigs.tsv\", \"w\") as outfile:\n",
    "    outfile.write(\"ncbi_taxid\\ttaxonomy\\n\")\n",
    "    for taxid in ncbi_taxid:\n",
    "        # print(taxid)\n",
    "        try:\n",
    "            handle = Entrez.efetch(db=\"Taxonomy\", id=taxid, retmode=\"xml\")\n",
    "            records = Entrez.read(handle)\n",
    "            ncbi_taxa_dict[taxid] = records[0]['Lineage']\n",
    "            writeout = str(taxid) + \"\\t\" + str(ncbi_taxa_dict[taxid]) +\"\\n\"\n",
    "            outfile.write(writeout)\n",
    "\n",
    "        except:\n",
    "            print(taxid)\n",
    "            writeout=str(taxid) + \"\\tNaN\\n\"\n",
    "            outfile.write(writeout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create ncbi taxid from tsv previously painstakingly generated using entrez\n",
    "ncbi_taxid_df = pd.read_csv(\n",
    "    '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/kaiju_out/ncbi_taxid_kaiju_contigs.tsv',\n",
    "           sep='\\t')\n",
    "# make dictionary to map taxid\n",
    "ncbi_taxa_dict = ncbi_taxid_df.set_index('ncbi_taxid').to_dict()['taxonomy']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 390,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "134537\n",
      "1620392\n",
      "2108470\n",
      "2855394\n",
      "32049\n",
      "1080068\n",
      "1116391\n",
      "232348\n",
      "321327\n",
      "1110509\n",
      "403\n",
      "1196095\n",
      "641\n",
      "2094119\n",
      "1752063\n",
      "2033869\n",
      "926570\n",
      "103690\n",
      "483547\n",
      "1006576\n",
      "702967\n",
      "586416\n",
      "1273155\n",
      "2814655\n",
      "1677857\n",
      "81479\n",
      "1826872\n",
      "2605945\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = 'alexa_nicolas@berkeley.edu'\n",
    "\n",
    "with open(\"/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/kaiju_out/ncbi_taxid_kaiju_contigs.tsv\", \"a\") as outfile:\n",
    "    for taxid in wsip_ncbi_taxid:\n",
    "        # print(taxid)\n",
    "        try:\n",
    "            print(taxid)\n",
    "            handle = Entrez.efetch(db=\"Taxonomy\", id=taxid, retmode=\"xml\")\n",
    "            records = Entrez.read(handle)\n",
    "            ncbi_taxa_dict[taxid] = records[0]['Lineage']\n",
    "            writeout = str(taxid) + \"\\t\" + str(ncbi_taxa_dict[taxid]) +\"\\n\"\n",
    "            outfile.write(writeout)\n",
    "\n",
    "        except:\n",
    "            print(taxid, np.nan)\n",
    "            writeout=str(taxid) + \"\\tNaN\\n\"\n",
    "            outfile.write(writeout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 662,
   "metadata": {},
   "outputs": [],
   "source": [
    "kaiju_out = pd.concat([kaiju_unbinned_spacers,kaiju_vOTUs])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "kaiju_out.to_csv(\n",
    "    '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/kaiju_out/kaiju_contig_taxonomy.csv',\n",
    "                index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 666,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove duplicates from Kaiju_out\n",
    "kaiju_out = kaiju_out.drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# vOTU spacers "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## make blast db & run BLAST to match to sequences from undereplicated MAGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/shared/software/bin/blastn -query /groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/4Wh2oV_minced_out/4Wh2oV_scaffold_spacers.fa -db /groups/banfield/projects/environmental/4thwedge/2020/assembly.d/all_MAGs_disrat0.33/concatenated_underep_MAGs/underep_MAGs.fa -num_threads 16 -outfmt ‘6 qseqid sseqid pident qcovs qstart qend sstart send length evalue bitscore mismatch’ -out /groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/4Wh2oV_minced_out/blast_out_underep_MAGs/underepMAGs_4Wh2Ovspacers.txt\n"
     ]
    }
   ],
   "source": [
    "virome_spacer_db = '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/4Wh2oV_minced_out/4Wh2oV_scaffold_spacers.fa'\n",
    "MAG_db = '/groups/banfield/projects/environmental/4thwedge/2020/assembly.d/all_MAGs_disrat0.33/concatenated_underep_MAGs/underep_MAGs.fa'\n",
    "blastn_cmd = '/shared/software/bin/blastn -query ' + virome_spacer_db + \" -db \" + \\\n",
    "                MAG_db + \" -num_threads 16 -outfmt ‘6 qseqid sseqid pident qcovs qstart qend sstart send length evalue bitscore mismatch’ -out \" + \\\n",
    "                \"/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/4Wh2oV_minced_out/blast_out_underep_MAGs/underepMAGs_4Wh2Ovspacers.txt\"\n",
    "print(blastn_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "vOTU_spacer_MAGs = pd.read_csv('/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/4Wh2oV_minced_out/blast_out_underep_MAGs/underepMAGs_4Wh2Ovspacers.txt',\n",
    "                                 header=None, sep='\\t',\n",
    "                                  names=['qseqid', 'sseqid', 'pident', 'qcovs', 'qstart', 'qend', 'sstart', 'send', 'length', 'evalue', 'bitscore', 'mismatch'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>mismatch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4Wh2oV_72h_scaffold_36124_CRISPR_15_spacer_6</td>\n",
       "      <td>12C16O_72h_100_unf_k141_138559</td>\n",
       "      <td>96.875</td>\n",
       "      <td>84</td>\n",
       "      <td>6</td>\n",
       "      <td>37</td>\n",
       "      <td>8282</td>\n",
       "      <td>8313</td>\n",
       "      <td>32</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>54.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         qseqid  \\\n",
       "0  4Wh2oV_72h_scaffold_36124_CRISPR_15_spacer_6   \n",
       "\n",
       "                           sseqid  pident  qcovs  qstart  qend  sstart  send  \\\n",
       "0  12C16O_72h_100_unf_k141_138559  96.875     84       6    37    8282  8313   \n",
       "\n",
       "   length    evalue  bitscore  mismatch  \n",
       "0      32  0.000002      54.7         1  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vOTU_spacer_MAGs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "only 1 vOTU spacer matched a MAG!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge MAG scaffold 2 bin file with MAG hit \n",
    "vOTU_spacer_MAGs = vOTU_spacer_MAGs.merge(s2b, \n",
    "                                          left_on='sseqid',right_on='scaffold',how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Additional matching tested\n",
    "\n",
    "We tested whether additional vOTU host taxonomy could be ascertained by looking for matches between vOTU spacers and unbinned metagenome contigs. We manually inspected matched unbinnned metagenome contigs on ggKbase and these contigs appeared to be viral (contained canonical viral genes) so we did not use unbinned metagenome contigs as another source of potential viral taxonomy. \n",
    "\n",
    "We also used BLAST to align vOTU spacers to IMG microbial and sequences (IMG-VR nucleotide sequences). Results were strange... only 1 vOTU spacer matched a sequence with a defined taxonomy and it looked like a virus that has a Clostridial host, but the vOTU is circularized and seems like a like proteobacteria/rhizobia-infecting virus https://ggkbase.berkeley.edu/4Wh2oV_168h/organisms/487641/contigs/493190824"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Spacer database (Shmakov et al. 2017)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -query /groups/banfield/users/alexa/db/spacers/Supplementary_material_spacers.fna -db /groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/4W_vOTU_db/viral_scaffold_set_12C16O_4Wh2oV.fa -num_threads 16 -outfmt ‘6 qseqid sseqid pident qcovs qstart qend sstart send length evalue bitscore mismatch’ -out /groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/4W_vOTU_set_out/4W_vOTU_set_spacerDB_blast_output_0-1mismatches.txt\n"
     ]
    }
   ],
   "source": [
    "spacer_db = '/groups/banfield/users/alexa/db/spacers/Supplementary_material_spacers.fna'\n",
    "vOTU_db = '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/4W_vOTU_db/viral_scaffold_set_12C16O_4Wh2oV.fa'\n",
    "outfile = \"/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/4W_vOTU_set_out/4W_vOTU_set_spacerDB_blast_output_0-1mismatches.txt\"\n",
    "blastn = \"blastn -query \" + spacer_db +  \" -db \" + vOTU_db + \\\n",
    "\" -num_threads 16 -outfmt ‘6 qseqid sseqid pident qcovs qstart qend sstart send length evalue bitscore mismatch’ -out \" + outfile\n",
    "print(blastn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "spacer_db_blast_result = pd.read_csv(\n",
    "    \"/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/4W_vOTU_set_out/4W_vOTU_set_spacerDB_blast_output_0-1mismatches.txt\",\n",
    "                                    sep='\\t',\n",
    "                                header=None,\n",
    "                                names = ['qseqid', 'sseqid', 'pident', 'qcovs', 'qstart', 'qend', 'sstart', 'send', 'length', 'evalue', 'bitscore', 'mismatch'])\n",
    "spacer_db_blast_result['ncbi_accession_spacer'] = spacer_db_blast_result['qseqid'].str.split(\"_\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make file that's just ncbi accessions\n",
    "with open(\n",
    "'/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/4W_vOTU_set_out/BLAST_hit_spacer_acccessions.txt','w') as filehandle:\n",
    "    filehandle.writelines(\"%s\\n\" % accession for accession in spacer_db_blast_result['ncbi_accession_spacer'].unique().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Bacteria', 'Proteobacteria', 'Betaproteobacteria', 'Burkholderiales', 'Comamonadaceae', 'Acidovorax']\n",
      "['Bacteria', 'Proteobacteria', 'Betaproteobacteria', 'Burkholderiales', 'Comamonadaceae', 'Lampropedia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Rhodanobacteraceae', 'Rhodanobacter', 'unclassified Rhodanobacter']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Rhodanobacteraceae', 'Rhodanobacter']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Rhodanobacteraceae', 'Rhodanobacter']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Rhodanobacteraceae', 'Rhodanobacter']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Rhodanobacteraceae', 'Rhodanobacter']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Rhodanobacteraceae', 'Rhodanobacter']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Actinobacteria', 'Corynebacteriales', 'Corynebacteriaceae', 'Corynebacterium']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Bacteroidetes', 'Bacteroidia', 'Bacteroidales']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Rhodobacterales', 'Rhodobacteraceae', 'Rhodobacter']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Rhodobacterales', 'Rhodobacteraceae', 'Rhodobacter']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pasteurellales', 'Pasteurellaceae', 'Bibersteinia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Rhodobacterales', 'Rhodobacteraceae']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Kitasatospora']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Kitasatospora']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Kitasatospora']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Xanthomonadales', 'Xanthomonadaceae', 'Xanthomonas']\n",
      "['Bacteria', 'Cyanobacteria', 'Nostocales', 'Calotrichaceae', 'Calothrix', 'unclassified Calothrix']\n",
      "['Bacteria', 'Actinobacteria', 'Micrococcales', 'Brevibacteriaceae', 'Brevibacterium']\n",
      "['Bacteria', 'Chlorobi', 'Chlorobia', 'Chlorobiales', 'Chlorobiaceae', 'Chlorobium/Pelodictyon group', 'Chlorobium']\n",
      "['Bacteria', 'Chloroflexi', 'Chloroflexia', 'Chloroflexales', 'Chloroflexineae', 'Chloroflexaceae', 'Chloroflexus']\n",
      "['Bacteria', 'Chloroflexi', 'Chloroflexia', 'Chloroflexales', 'Chloroflexineae', 'Chloroflexaceae', 'Chloroflexus']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Micrococcales', 'Promicromonosporaceae', 'Xylanimonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Betaproteobacteria', 'Burkholderiales', 'Burkholderiaceae', 'Burkholderia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Sphingomonadales', 'Sphingomonadaceae', 'Sphingomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Sphingomonadales', 'Sphingomonadaceae', 'Sphingomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Sphingomonadales', 'Sphingomonadaceae', 'Sphingomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Sphingomonadales', 'Sphingomonadaceae', 'Sphingomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Sphingomonadales', 'Sphingomonadaceae', 'Sphingomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Sphingomonadales', 'Sphingomonadaceae', 'Sphingomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Alphaproteobacteria', 'Rhodobacterales', 'Roseobacteraceae', 'Rubellimicrobium']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Klebsiella']\n",
      "['Bacteria', 'Firmicutes', 'Bacilli', 'Bacillales', 'Bacillaceae', 'Parageobacillus']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Klebsiella']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Archaea', 'Euryarchaeota', 'Stenosarchaea group', 'Methanomicrobia', 'Methanomicrobiales', 'Methanomicrobiaceae', 'Methanoculleus']\n",
      "['Archaea', 'Euryarchaeota', 'Stenosarchaea group', 'Methanomicrobia', 'Methanomicrobiales', 'Methanomicrobiaceae', 'Methanoculleus']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Deltaproteobacteria', 'Desulfobacterales', 'Desulfobacteraceae', 'Desulfatitalea']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Salmonella']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Salmonella']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Escherichia']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Salmonella']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Salmonella']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "['Bacteria', 'Actinobacteria', 'Streptomycetales', 'Streptomycetaceae', 'Streptomyces']\n",
      "CP006567.1\n"
     ]
    }
   ],
   "source": [
    "# takes a long time to run through notebook, can copy code into terminal running python\n",
    "ncbi_accession_spacers = []\n",
    "with open('/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/4W_vOTU_set_out/BLAST_hit_spacer_acccessions.txt','r') as file:\n",
    "    for line in file.readlines():\n",
    "        ncbi_accession_spacers.append(line.strip())\n",
    "\n",
    "accession_tax_dict = {}\n",
    "with open(\"/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/4W_vOTU_set_out/spacersDB_blast_out_retrieve_taxa.tsv\", \"w\") as outfile:\n",
    "    outfile.write(\"ncbi_accession_spacer\\ttaxonomy\\torganism\\n\")\n",
    "    for accession in ncbi_accession_spacers:\n",
    "#         print(accession)\n",
    "        try:\n",
    "            handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"gb\", retmode=\"text\")\n",
    "            records = SeqIO.parse(handle, \"genbank\")\n",
    "            for record in records:\n",
    "                accession_tax_dict[accession] = [record.annotations['taxonomy'],record.annotations['organism']]\n",
    "                writeout = accession + \"\\t\" + str(record.annotations['taxonomy']) + \"\\t\" + str(record.annotations['organism']+\"\\n\")\n",
    "                outfile.write(writeout)\n",
    "                print(record.annotations['taxonomy'])\n",
    "\n",
    "        except:\n",
    "            print(accession)\n",
    "            writeout=accession + \"\\tbreak\\tbreak\\n\"\n",
    "            outfile.write(writeout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_accession_spacers = pd.read_csv(\"/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/4W_vOTU_set_out/spacersDB_blast_out_retrieve_taxa.tsv\",\n",
    "                                     sep='\\t').replace('break',np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CP007574.1\n"
     ]
    }
   ],
   "source": [
    "accession_tax_dict = {}\n",
    "\n",
    "for accession in ncbi_accession_spacers[ncbi_accession_spacers['organism']=='break']['ncbi_accession_spacer'].to_list():\n",
    "#         print(accession)\n",
    "    try:\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"gb\", retmode=\"text\")\n",
    "        records = SeqIO.parse(handle, \"genbank\")\n",
    "        for record in records:\n",
    "            accession_tax_dict[accession] = [record.annotations['taxonomy'],record.annotations['organism']]\n",
    "#             writeout = accession + \"\\t\" + str(record.annotations['taxonomy']) + \"\\t\" + str(record.annotations['organism']+\"\\n\")\n",
    "#             outfile.write(writeout)\n",
    "#             print(record.annotations['taxonomy'])\n",
    "\n",
    "    except:\n",
    "        print(accession)\n",
    "#         writeout=accession + \"\\tbreak\\tbreak\\n\"\n",
    "#         outfile.write(writeout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 481,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hardcoding last accession number...\n",
    "accession = 'CP007574.1'\n",
    "try:\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"gb\", retmode=\"text\")\n",
    "    records = SeqIO.parse(handle, \"genbank\")\n",
    "    for record in records:\n",
    "        accession_tax_dict[accession] = [record.annotations['taxonomy'],record.annotations['organism']]\n",
    "except:\n",
    "    print(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_accession_spacers['ncbi_spacer_phyla'] = ncbi_accession_spacers['taxonomy'].str.split(\",\").str[1].str.strip(\",' '\")\n",
    "ncbi_accession_spacers['ncbi_spacer_genera'] = ncbi_accession_spacers['taxonomy'].str.rsplit(\",\",-1).str[-1].str.strip(\",' ']\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 551,
   "metadata": {},
   "outputs": [],
   "source": [
    "fix_taxa = pd.DataFrame.from_dict(accession_tax_dict).T.reset_index().rename(columns={'index':'ncbi_accession_spacer',\n",
    "                                                                           0:'taxonomy',1:'organism'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 649,
   "metadata": {},
   "outputs": [],
   "source": [
    "fix_taxa['ncbi_spacer_phyla'] = fix_taxa['taxonomy'].str[1]\n",
    "fix_taxa['ncbi_spacer_genera'] = fix_taxa['taxonomy'].str[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 653,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use combine_first to add in taxa from fix_taxa \n",
    "ncbi_accession_spacers_taxa = ncbi_accession_spacers.combine_first(ncbi_accession_spacers[['ncbi_accession_spacer']].merge(fix_taxa,how='left'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 661,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_accession_spacers_taxa.to_csv(\n",
    "    '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/4W_vOTU_set_out/4W_vOTU_spacersDB_blastout_taxa.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ncbi spacer accession df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_accession_spacers_taxa = pd.read_csv(\n",
    "'/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/4W_vOTU_set_out/4W_vOTU_spacersDB_blastout_taxa.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "merge spacer taxonomy back with viral hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "spacer_db_blast_result = spacer_db_blast_result.merge(ncbi_accession_spacers_taxa, on='ncbi_accession_spacer',how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>ncbi_accession_spacer</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>organism</th>\n",
       "      <th>ncbi_spacer_phyla</th>\n",
       "      <th>ncbi_spacer_genera</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP002521.1_300343_306222_45_spacer_303283_30</td>\n",
       "      <td>4Wh2oV_72h_scaffold_299905</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>5038</td>\n",
       "      <td>5067</td>\n",
       "      <td>30</td>\n",
       "      <td>4.340000e-08</td>\n",
       "      <td>56.5</td>\n",
       "      <td>0</td>\n",
       "      <td>CP002521.1</td>\n",
       "      <td>['Bacteria', 'Proteobacteria', 'Betaproteobact...</td>\n",
       "      <td>Acidovorax avenae subsp. avenae ATCC 19860</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Acidovorax</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         qseqid                      sseqid  \\\n",
       "0  CP002521.1_300343_306222_45_spacer_303283_30  4Wh2oV_72h_scaffold_299905   \n",
       "\n",
       "   pident  qcovs  qstart  qend  sstart  send  length        evalue  bitscore  \\\n",
       "0   100.0    100       1    30    5038  5067      30  4.340000e-08      56.5   \n",
       "\n",
       "   mismatch ncbi_accession_spacer  \\\n",
       "0         0            CP002521.1   \n",
       "\n",
       "                                            taxonomy  \\\n",
       "0  ['Bacteria', 'Proteobacteria', 'Betaproteobact...   \n",
       "\n",
       "                                     organism ncbi_spacer_phyla  \\\n",
       "0  Acidovorax avenae subsp. avenae ATCC 19860    Proteobacteria   \n",
       "\n",
       "  ncbi_spacer_genera  \n",
       "0         Acidovorax  "
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spacer_db_blast_result.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_spacer_match = virus_scaff_set_drep.merge(spacer_db_blast_result.rename(columns={'length':'length_spacer_match'}), left_on='scaffold',right_on=\"sseqid\",how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop duplicate spacer match to same strain organism\n",
    "ncbi_spacer_match_dd = ncbi_spacer_match.drop_duplicates(subset=['scaffold','taxonomy','organism'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_spacer_match_dd.to_csv(\"/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/annotations.d/spacer_db_out/spacer_db_viral_contig.csv\",index=False\n",
    "                           )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>v_sum</th>\n",
       "      <th>bin</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>organism</th>\n",
       "      <th>organism taxonomy</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>GC %</th>\n",
       "      <th>Taxonomy winner</th>\n",
       "      <th>Winner %</th>\n",
       "      <th>...</th>\n",
       "      <th>pident</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>length_spacer_match</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>ncbi_accession_spacer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4Wh2oV_0h_scaffold_10786</td>\n",
       "      <td>3</td>\n",
       "      <td>4Wh2oV_0h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces avermitilis MA-4680 = NBRC 14893</td>\n",
       "      <td>unknown</td>\n",
       "      <td>136.15</td>\n",
       "      <td>68.82</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>0.64</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>15971</td>\n",
       "      <td>16002</td>\n",
       "      <td>32</td>\n",
       "      <td>2.120000e-08</td>\n",
       "      <td>60.2</td>\n",
       "      <td>JZJK01000038.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4Wh2oV_0h_scaffold_10786</td>\n",
       "      <td>3</td>\n",
       "      <td>4Wh2oV_0h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces avermitilis</td>\n",
       "      <td>unknown</td>\n",
       "      <td>136.15</td>\n",
       "      <td>68.82</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>0.64</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>16002</td>\n",
       "      <td>15971</td>\n",
       "      <td>32</td>\n",
       "      <td>2.120000e-08</td>\n",
       "      <td>60.2</td>\n",
       "      <td>KQ948608.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4Wh2oV_0h_scaffold_1114760</td>\n",
       "      <td>3</td>\n",
       "      <td>4Wh2oV_0h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces sp. Root1295</td>\n",
       "      <td>unknown</td>\n",
       "      <td>50.77</td>\n",
       "      <td>69.35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>97</td>\n",
       "      <td>1</td>\n",
       "      <td>31</td>\n",
       "      <td>10442</td>\n",
       "      <td>10412</td>\n",
       "      <td>31</td>\n",
       "      <td>7.640000e-08</td>\n",
       "      <td>58.4</td>\n",
       "      <td>LMEL01000024.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4Wh2oV_0h_scaffold_1114760</td>\n",
       "      <td>3</td>\n",
       "      <td>4Wh2oV_0h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces sp. Root63</td>\n",
       "      <td>unknown</td>\n",
       "      <td>50.77</td>\n",
       "      <td>69.35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>97</td>\n",
       "      <td>1</td>\n",
       "      <td>31</td>\n",
       "      <td>10442</td>\n",
       "      <td>10412</td>\n",
       "      <td>31</td>\n",
       "      <td>7.640000e-08</td>\n",
       "      <td>58.4</td>\n",
       "      <td>LMGX01000007.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4Wh2oV_0h_scaffold_1114760</td>\n",
       "      <td>3</td>\n",
       "      <td>4Wh2oV_0h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces avermitilis</td>\n",
       "      <td>unknown</td>\n",
       "      <td>50.77</td>\n",
       "      <td>69.35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>44443</td>\n",
       "      <td>44412</td>\n",
       "      <td>32</td>\n",
       "      <td>2.120000e-08</td>\n",
       "      <td>60.2</td>\n",
       "      <td>KQ948608.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1228</th>\n",
       "      <td>4Wh2oV_168h_scaffold_491050</td>\n",
       "      <td>4</td>\n",
       "      <td>4Wh2oV_168h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces bikiniensis</td>\n",
       "      <td>unknown</td>\n",
       "      <td>757.45</td>\n",
       "      <td>70.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>96.97</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>6706</td>\n",
       "      <td>6738</td>\n",
       "      <td>33</td>\n",
       "      <td>3.140000e-07</td>\n",
       "      <td>56.5</td>\n",
       "      <td>JNWL01000041.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1231</th>\n",
       "      <td>4Wh2oV_168h_scaffold_491050</td>\n",
       "      <td>4</td>\n",
       "      <td>4Wh2oV_168h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces sp. NRRL F-5527</td>\n",
       "      <td>unknown</td>\n",
       "      <td>757.45</td>\n",
       "      <td>70.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>71979</td>\n",
       "      <td>71947</td>\n",
       "      <td>33</td>\n",
       "      <td>6.750000e-09</td>\n",
       "      <td>62.1</td>\n",
       "      <td>JOHL01000066.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1233</th>\n",
       "      <td>4Wh2oV_168h_scaffold_491050</td>\n",
       "      <td>4</td>\n",
       "      <td>4Wh2oV_168h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces sp. NRRL S-1314</td>\n",
       "      <td>unknown</td>\n",
       "      <td>757.45</td>\n",
       "      <td>70.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>100.00</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>71947</td>\n",
       "      <td>71979</td>\n",
       "      <td>33</td>\n",
       "      <td>6.750000e-09</td>\n",
       "      <td>62.1</td>\n",
       "      <td>JOHU01000052.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1235</th>\n",
       "      <td>4Wh2oV_168h_scaffold_546933</td>\n",
       "      <td>4</td>\n",
       "      <td>4Wh2oV_168h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces sp. NRRL S-146</td>\n",
       "      <td>unknown</td>\n",
       "      <td>734.61</td>\n",
       "      <td>68.11</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>0.60</td>\n",
       "      <td>...</td>\n",
       "      <td>96.97</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>52425</td>\n",
       "      <td>52457</td>\n",
       "      <td>33</td>\n",
       "      <td>3.140000e-07</td>\n",
       "      <td>56.5</td>\n",
       "      <td>JOAW01000503.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236</th>\n",
       "      <td>4Wh2oV_168h_scaffold_546933</td>\n",
       "      <td>4</td>\n",
       "      <td>4Wh2oV_168h_UNK</td>\n",
       "      <td>['Bacteria', 'Actinobacteria', 'Streptomycetal...</td>\n",
       "      <td>Streptomyces sp. PsTaAH-124</td>\n",
       "      <td>unknown</td>\n",
       "      <td>734.61</td>\n",
       "      <td>68.11</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>0.60</td>\n",
       "      <td>...</td>\n",
       "      <td>96.97</td>\n",
       "      <td>97</td>\n",
       "      <td>2</td>\n",
       "      <td>33</td>\n",
       "      <td>7369</td>\n",
       "      <td>7401</td>\n",
       "      <td>33</td>\n",
       "      <td>1.130000e-06</td>\n",
       "      <td>54.7</td>\n",
       "      <td>KB892181.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>652 rows × 101 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         scaffold  v_sum              bin  \\\n",
       "4        4Wh2oV_0h_scaffold_10786      3    4Wh2oV_0h_UNK   \n",
       "7        4Wh2oV_0h_scaffold_10786      3    4Wh2oV_0h_UNK   \n",
       "9      4Wh2oV_0h_scaffold_1114760      3    4Wh2oV_0h_UNK   \n",
       "10     4Wh2oV_0h_scaffold_1114760      3    4Wh2oV_0h_UNK   \n",
       "11     4Wh2oV_0h_scaffold_1114760      3    4Wh2oV_0h_UNK   \n",
       "...                           ...    ...              ...   \n",
       "1228  4Wh2oV_168h_scaffold_491050      4  4Wh2oV_168h_UNK   \n",
       "1231  4Wh2oV_168h_scaffold_491050      4  4Wh2oV_168h_UNK   \n",
       "1233  4Wh2oV_168h_scaffold_491050      4  4Wh2oV_168h_UNK   \n",
       "1235  4Wh2oV_168h_scaffold_546933      4  4Wh2oV_168h_UNK   \n",
       "1236  4Wh2oV_168h_scaffold_546933      4  4Wh2oV_168h_UNK   \n",
       "\n",
       "                                               taxonomy  \\\n",
       "4     ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "7     ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "9     ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "10    ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "11    ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "...                                                 ...   \n",
       "1228  ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "1231  ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "1233  ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "1235  ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "1236  ['Bacteria', 'Actinobacteria', 'Streptomycetal...   \n",
       "\n",
       "                                           organism organism taxonomy  \\\n",
       "4     Streptomyces avermitilis MA-4680 = NBRC 14893           unknown   \n",
       "7                          Streptomyces avermitilis           unknown   \n",
       "9                         Streptomyces sp. Root1295           unknown   \n",
       "10                          Streptomyces sp. Root63           unknown   \n",
       "11                         Streptomyces avermitilis           unknown   \n",
       "...                                             ...               ...   \n",
       "1228                       Streptomyces bikiniensis           unknown   \n",
       "1231                   Streptomyces sp. NRRL F-5527           unknown   \n",
       "1233                   Streptomyces sp. NRRL S-1314           unknown   \n",
       "1235                    Streptomyces sp. NRRL S-146           unknown   \n",
       "1236                    Streptomyces sp. PsTaAH-124           unknown   \n",
       "\n",
       "      Coverage   GC % Taxonomy winner  Winner %  ...  pident  qcovs qstart  \\\n",
       "4       136.15  68.82    Streptomyces      0.64  ...  100.00    100      1   \n",
       "7       136.15  68.82    Streptomyces      0.64  ...  100.00    100      1   \n",
       "9        50.77  69.35             NaN       NaN  ...  100.00     97      1   \n",
       "10       50.77  69.35             NaN       NaN  ...  100.00     97      1   \n",
       "11       50.77  69.35             NaN       NaN  ...  100.00    100      1   \n",
       "...        ...    ...             ...       ...  ...     ...    ...    ...   \n",
       "1228    757.45  70.91             NaN       NaN  ...   96.97    100      1   \n",
       "1231    757.45  70.91             NaN       NaN  ...  100.00    100      1   \n",
       "1233    757.45  70.91             NaN       NaN  ...  100.00    100      1   \n",
       "1235    734.61  68.11    Streptomyces      0.60  ...   96.97    100      1   \n",
       "1236    734.61  68.11    Streptomyces      0.60  ...   96.97     97      2   \n",
       "\n",
       "      qend sstart   send length_spacer_match        evalue bitscore  \\\n",
       "4       32  15971  16002                  32  2.120000e-08     60.2   \n",
       "7       32  16002  15971                  32  2.120000e-08     60.2   \n",
       "9       31  10442  10412                  31  7.640000e-08     58.4   \n",
       "10      31  10442  10412                  31  7.640000e-08     58.4   \n",
       "11      32  44443  44412                  32  2.120000e-08     60.2   \n",
       "...    ...    ...    ...                 ...           ...      ...   \n",
       "1228    33   6706   6738                  33  3.140000e-07     56.5   \n",
       "1231    33  71979  71947                  33  6.750000e-09     62.1   \n",
       "1233    33  71947  71979                  33  6.750000e-09     62.1   \n",
       "1235    33  52425  52457                  33  3.140000e-07     56.5   \n",
       "1236    33   7369   7401                  33  1.130000e-06     54.7   \n",
       "\n",
       "      ncbi_accession_spacer  \n",
       "4            JZJK01000038.1  \n",
       "7                KQ948608.1  \n",
       "9            LMEL01000024.1  \n",
       "10           LMGX01000007.1  \n",
       "11               KQ948608.1  \n",
       "...                     ...  \n",
       "1228         JNWL01000041.1  \n",
       "1231         JOHL01000066.1  \n",
       "1233         JOHU01000052.1  \n",
       "1235         JOAW01000503.1  \n",
       "1236             KB892181.1  \n",
       "\n",
       "[652 rows x 101 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_spacer_match_dd[ncbi_spacer_match_dd.duplicated(subset='scaffold',keep=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Use Spacers from metagenome assemblies to connect host and virus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spacers from MAGs and blast vOTU hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'blastn -query /groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/MAGs_nonderep_spacers_wbinname.fa -db /groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/4W_vOTU_db/viral_scaffold_set_12C16O_4Wh2oV.fa -num_threads 16 -outfmt ‘6 qseqid sseqid pident qcovs qstart qend sstart send length evalue bitscore mismatch’ -out /groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/MAGs_nonderep_spacers_4WvOTUs.tsv'"
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mag_spacers = '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/MAGs_nonderep_spacers_wbinname.fa'\n",
    "out = '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/MAGs_nonderep_spacers_4WvOTUs.tsv'\n",
    "blastn = \"blastn -query \" + mag_spacers +  \" -db \" + vOTU_db + \\\n",
    "\" -num_threads 16 -outfmt ‘6 qseqid sseqid pident qcovs qstart qend sstart send length evalue bitscore mismatch’ -out \" + out\n",
    "blastn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "mag_spacer_hits = pd.read_csv('/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/MAGs_nonderep_spacers_4WvOTUs.tsv',\n",
    "                             sep='\\t',\n",
    "                             header=None,\n",
    "                             names = [\n",
    "                                'qseqid', 'sseqid', 'pident', 'qcovs', 'qstart', 'qend', 'sstart', 'send', 'length', 'evalue', 'bitscore', 'mismatch' \n",
    "                             ])\n",
    "mag_spacer_hits['scaffold'] = mag_spacer_hits['qseqid'].str.split(\"@\").str[1].str.rsplit(\"_\",1).str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "mag_spacer_hits = pd.merge(mag_spacer_hits,s2b,on='scaffold',how='left')#.sort_values('classification',ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "mag_spacer_hits['votu'] = mag_spacer_hits['sseqid']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## unbinned MG contigs with spacer arrays kaiju taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 709,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_dir = '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/'\n",
    "mg_spacers = crispr_dir+'4W_contigs_spacers.fa'\n",
    "# read in blast output for mg unbinned crispr arrays aligning with vOTU sequence \n",
    "mg_spacers_blastout = pd.read_csv('/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/4W_contigs_spacers_vs_4Wh2oV_blast_db.tsv',\n",
    "                                  sep='\\t')\n",
    "mg_spacers_blastout = mg_spacers_blastout['qseqid'].str.split(\"_CRISPR\",expand=True).rename(columns={0:'mg_contig',1:'array'}).merge(mg_spacers_blastout,left_index=True,right_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "multiple spacers hit multiple places in votu from single contig, keep this type of duplication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 713,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge unbinned spacer blast from mg with kaiju taxonomy\n",
    "# also issue with kaiju_out – has multiple entries per mg_contig ... \n",
    "mg_spacers_blastout_taxa = mg_spacers_blastout.merge(\n",
    "    kaiju_out[kaiju_out['kaiju_scaffold'].notna()],\n",
    "                                                     left_on='mg_contig',\n",
    "                                                     right_on='kaiju_scaffold', \n",
    "                                                     how='left',validate='m:1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3 contigs do not have assigned taxonomy from kaiju – these are unclassified\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "wsip_mag_file = '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/wSIP_MAGs_nonderep_spacers_vs_4Wh2oV_blast_db.tsv'\n",
    "wsip_mag = pd.read_csv(wsip_mag_file,\n",
    "                       sep='\\t')\n",
    "wsip_mag.rename(columns={'e-value':'evalue','pid':'pident','qcov':'qcovs'},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "wsip_mag['MAG'] = wsip_mag['MAG'].str.replace(\"_spacers.fa\",\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1-18-all-fractions_k255_4602126_CRISPR_1_spac...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H1-18-all-fractions_k255_4902837_CRISPR_1_spac...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H3-18-all-fractions_k255_4574136_CRISPR_1_spac...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spac...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spac...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spac...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spac...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spac...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              qseqid  sseqid\n",
       "0  H1-18-all-fractions_k255_4602126_CRISPR_1_spac...       5\n",
       "1  H1-18-all-fractions_k255_4902837_CRISPR_1_spac...       1\n",
       "2  H3-18-all-fractions_k255_4574136_CRISPR_1_spac...       1\n",
       "3  S3-16-all-fractions_k255_1197418_CRISPR_1_spac...       1\n",
       "4  S3-16-all-fractions_k255_1197418_CRISPR_1_spac...       2\n",
       "5  S3-16-all-fractions_k255_1197418_CRISPR_1_spac...       3\n",
       "6  S3-16-all-fractions_k255_1197418_CRISPR_1_spac...       3\n",
       "7  S3-16-all-fractions_k255_1197418_CRISPR_1_spac...       1"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wsip_mag.groupby(['qseqid'],as_index=False)['sseqid'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "wsip_mag['scaffold_id'] = wsip_mag['qseqid'].str.split(\"_C\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sseqid</th>\n",
       "      <th>qseqid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4Wh2oV_0h_scaffold_1703601</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4Wh2oV_0h_scaffold_637225</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4Wh2oV_24h_scaffold_1255333</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4Wh2oV_24h_scaffold_1500359</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4Wh2oV_24h_scaffold_1626649</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4Wh2oV_24h_scaffold_1787552</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4Wh2oV_24h_scaffold_221362</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4Wh2oV_24h_scaffold_467538</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4Wh2oV_24h_scaffold_960306</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4Wh2oV_3h_scaffold_1562313</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4Wh2oV_3h_scaffold_1843257</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4Wh2oV_3h_scaffold_1894264</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4Wh2oV_3h_scaffold_356956</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>4Wh2oV_48h_scaffold_716693</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4Wh2oV_48h_scaffold_936400</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4Wh2oV_72h_scaffold_529029</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>4Wh2oV_72h_scaffold_695820</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         sseqid  qseqid\n",
       "0    4Wh2oV_0h_scaffold_1703601       1\n",
       "1     4Wh2oV_0h_scaffold_637225       1\n",
       "2   4Wh2oV_24h_scaffold_1255333       1\n",
       "3   4Wh2oV_24h_scaffold_1500359       1\n",
       "4   4Wh2oV_24h_scaffold_1626649       1\n",
       "5   4Wh2oV_24h_scaffold_1787552       1\n",
       "6    4Wh2oV_24h_scaffold_221362       1\n",
       "7    4Wh2oV_24h_scaffold_467538       1\n",
       "8    4Wh2oV_24h_scaffold_960306       1\n",
       "9    4Wh2oV_3h_scaffold_1562313       1\n",
       "10   4Wh2oV_3h_scaffold_1843257       1\n",
       "11   4Wh2oV_3h_scaffold_1894264       1\n",
       "12    4Wh2oV_3h_scaffold_356956       1\n",
       "13   4Wh2oV_48h_scaffold_716693       1\n",
       "14   4Wh2oV_48h_scaffold_936400       1\n",
       "15   4Wh2oV_72h_scaffold_529029       1\n",
       "16   4Wh2oV_72h_scaffold_695820       1"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wsip_mag.groupby(['sseqid'],as_index=False)['qseqid'].count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "merge scaffold to bin with wsip mag hits – not working check in w Ella"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>Bin</th>\n",
       "      <th>classification</th>\n",
       "      <th>scaffold_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [scaffold, Bin, classification, scaffold_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 417,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2b[s2b['scaffold_id'].isin(wsip_mag['scaffold_id'].unique())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>Bin</th>\n",
       "      <th>classification</th>\n",
       "      <th>bin</th>\n",
       "      <th>secondary_cluster</th>\n",
       "      <th>NCBI_taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>698817</th>\n",
       "      <td>H3-18-all-fractions_k255_2339 read_length_150 read_count_31</td>\n",
       "      <td>H3-18-all-fractions_conc_114.Actinobacteria_73_10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H3-18-all-fractions_conc_114_Actinobacteria_73_10</td>\n",
       "      <td>122_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698818</th>\n",
       "      <td>H3-18-all-fractions_k255_3194 read_length_150 read_count_115</td>\n",
       "      <td>H3-18-all-fractions_conc_114.Actinobacteria_73_10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H3-18-all-fractions_conc_114_Actinobacteria_73_10</td>\n",
       "      <td>122_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698819</th>\n",
       "      <td>H3-18-all-fractions_k255_4549 read_length_150 read_count_21</td>\n",
       "      <td>H3-18-all-fractions_conc_114.Actinobacteria_73_10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H3-18-all-fractions_conc_114_Actinobacteria_73_10</td>\n",
       "      <td>122_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698820</th>\n",
       "      <td>H3-18-all-fractions_k255_6336 read_length_150 read_count_19</td>\n",
       "      <td>H3-18-all-fractions_conc_114.Actinobacteria_73_10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H3-18-all-fractions_conc_114_Actinobacteria_73_10</td>\n",
       "      <td>122_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>698821</th>\n",
       "      <td>H3-18-all-fractions_k255_20509 read_length_150 read_count_17</td>\n",
       "      <td>H3-18-all-fractions_conc_114.Actinobacteria_73_10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H3-18-all-fractions_conc_114_Actinobacteria_73_10</td>\n",
       "      <td>122_1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701133</th>\n",
       "      <td>H3-18-all-fractions_k255_6607254 read_length_150 read_count_27</td>\n",
       "      <td>H3-18-all-fractions_conc_15.Actinobacteria_69_13</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__</td>\n",
       "      <td>H3-18-all-fractions_conc_15_Actinobacteria_69_13</td>\n",
       "      <td>163_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701134</th>\n",
       "      <td>H3-18-all-fractions_k255_6608650 read_length_150 read_count_31</td>\n",
       "      <td>H3-18-all-fractions_conc_15.Actinobacteria_69_13</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__</td>\n",
       "      <td>H3-18-all-fractions_conc_15_Actinobacteria_69_13</td>\n",
       "      <td>163_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701135</th>\n",
       "      <td>H3-18-all-fractions_k255_6609022 read_length_150 read_count_63</td>\n",
       "      <td>H3-18-all-fractions_conc_15.Actinobacteria_69_13</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__</td>\n",
       "      <td>H3-18-all-fractions_conc_15_Actinobacteria_69_13</td>\n",
       "      <td>163_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701136</th>\n",
       "      <td>H3-18-all-fractions_k255_6610382 read_length_150 read_count_51</td>\n",
       "      <td>H3-18-all-fractions_conc_15.Actinobacteria_69_13</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__</td>\n",
       "      <td>H3-18-all-fractions_conc_15_Actinobacteria_69_13</td>\n",
       "      <td>163_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701137</th>\n",
       "      <td>H3-18-all-fractions_k255_6619471 read_length_150 read_count_20</td>\n",
       "      <td>H3-18-all-fractions_conc_15.Actinobacteria_69_13</td>\n",
       "      <td>d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__</td>\n",
       "      <td>H3-18-all-fractions_conc_15_Actinobacteria_69_13</td>\n",
       "      <td>163_0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2321 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                              scaffold  \\\n",
       "698817     H3-18-all-fractions_k255_2339 read_length_150 read_count_31   \n",
       "698818    H3-18-all-fractions_k255_3194 read_length_150 read_count_115   \n",
       "698819     H3-18-all-fractions_k255_4549 read_length_150 read_count_21   \n",
       "698820     H3-18-all-fractions_k255_6336 read_length_150 read_count_19   \n",
       "698821    H3-18-all-fractions_k255_20509 read_length_150 read_count_17   \n",
       "...                                                                ...   \n",
       "701133  H3-18-all-fractions_k255_6607254 read_length_150 read_count_27   \n",
       "701134  H3-18-all-fractions_k255_6608650 read_length_150 read_count_31   \n",
       "701135  H3-18-all-fractions_k255_6609022 read_length_150 read_count_63   \n",
       "701136  H3-18-all-fractions_k255_6610382 read_length_150 read_count_51   \n",
       "701137  H3-18-all-fractions_k255_6619471 read_length_150 read_count_20   \n",
       "\n",
       "                                                      Bin  \\\n",
       "698817  H3-18-all-fractions_conc_114.Actinobacteria_73_10   \n",
       "698818  H3-18-all-fractions_conc_114.Actinobacteria_73_10   \n",
       "698819  H3-18-all-fractions_conc_114.Actinobacteria_73_10   \n",
       "698820  H3-18-all-fractions_conc_114.Actinobacteria_73_10   \n",
       "698821  H3-18-all-fractions_conc_114.Actinobacteria_73_10   \n",
       "...                                                   ...   \n",
       "701133   H3-18-all-fractions_conc_15.Actinobacteria_69_13   \n",
       "701134   H3-18-all-fractions_conc_15.Actinobacteria_69_13   \n",
       "701135   H3-18-all-fractions_conc_15.Actinobacteria_69_13   \n",
       "701136   H3-18-all-fractions_conc_15.Actinobacteria_69_13   \n",
       "701137   H3-18-all-fractions_conc_15.Actinobacteria_69_13   \n",
       "\n",
       "                                                                                                            classification  \\\n",
       "698817                                                                                                                 NaN   \n",
       "698818                                                                                                                 NaN   \n",
       "698819                                                                                                                 NaN   \n",
       "698820                                                                                                                 NaN   \n",
       "698821                                                                                                                 NaN   \n",
       "...                                                                                                                    ...   \n",
       "701133  d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__   \n",
       "701134  d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__   \n",
       "701135  d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__   \n",
       "701136  d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__   \n",
       "701137  d__Bacteria;p__Actinobacteriota;c__Thermoleophilia;o__Solirubrobacterales;f__Solirubrobacteraceae;g__Palsa-465;s__   \n",
       "\n",
       "                                                      bin secondary_cluster  \\\n",
       "698817  H3-18-all-fractions_conc_114_Actinobacteria_73_10             122_1   \n",
       "698818  H3-18-all-fractions_conc_114_Actinobacteria_73_10             122_1   \n",
       "698819  H3-18-all-fractions_conc_114_Actinobacteria_73_10             122_1   \n",
       "698820  H3-18-all-fractions_conc_114_Actinobacteria_73_10             122_1   \n",
       "698821  H3-18-all-fractions_conc_114_Actinobacteria_73_10             122_1   \n",
       "...                                                   ...               ...   \n",
       "701133   H3-18-all-fractions_conc_15_Actinobacteria_69_13             163_0   \n",
       "701134   H3-18-all-fractions_conc_15_Actinobacteria_69_13             163_0   \n",
       "701135   H3-18-all-fractions_conc_15_Actinobacteria_69_13             163_0   \n",
       "701136   H3-18-all-fractions_conc_15_Actinobacteria_69_13             163_0   \n",
       "701137   H3-18-all-fractions_conc_15_Actinobacteria_69_13             163_0   \n",
       "\n",
       "       NCBI_taxonomy  \n",
       "698817           NaN  \n",
       "698818           NaN  \n",
       "698819           NaN  \n",
       "698820           NaN  \n",
       "698821           NaN  \n",
       "...              ...  \n",
       "701133           NaN  \n",
       "701134           NaN  \n",
       "701135           NaN  \n",
       "701136           NaN  \n",
       "701137           NaN  \n",
       "\n",
       "[2321 rows x 6 columns]"
      ]
     },
     "execution_count": 401,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2b[s2b['Bin'].str.contains('H3-18-all-fractions_conc_')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 676,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>classified</th>\n",
       "      <th>kaiju_scaffold</th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>kaiju_taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>C</td>\n",
       "      <td>H3-18-all-fractions_k255_4574136</td>\n",
       "      <td>983917</td>\n",
       "      <td>cellular organisms; Bacteria; Proteobacteria; Betaproteobacteria; Burkholderiales; Burkholderiales genera incertae sedis; Rubrivivax; Rubrivivax gelatinosus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    classified                    kaiju_scaffold  ncbi_taxid  \\\n",
       "170          C  H3-18-all-fractions_k255_4574136      983917   \n",
       "\n",
       "                                                                                                                                                   kaiju_taxonomy  \n",
       "170  cellular organisms; Bacteria; Proteobacteria; Betaproteobacteria; Burkholderiales; Burkholderiales genera incertae sedis; Rubrivivax; Rubrivivax gelatinosus  "
      ]
     },
     "execution_count": 676,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kaiju_wsip[kaiju_wsip['kaiju_scaffold'].isin(wsip_mag['scaffold_id'].unique())]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "this contig actually belongs to a MAG that should have an assignment – H3-18-all-fractions_Burkholderiales_71_8\n",
    "or H3-18-all-fractions_conc_96\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 421,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['H1-18-all-fractions_k255_4602126',\n",
       "       'H1-18-all-fractions_k255_4902837',\n",
       "       'H3-18-all-fractions_k255_4574136',\n",
       "       'S3-16-all-fractions_k255_1197418'], dtype=object)"
      ]
     },
     "execution_count": 421,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wsip_mag['scaffold_id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>Bin</th>\n",
       "      <th>classification</th>\n",
       "      <th>scaffold_id</th>\n",
       "      <th>secondary_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [scaffold, Bin, classification, scaffold_id, secondary_cluster]\n",
       "Index: []"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2b[s2b['scaffold_id'].isin(wsip_mag['scaffold_id'].unique())]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "why does wSIP MAG contigs not show up in s2b..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>length</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>MAG</th>\n",
       "      <th>scaffold_id</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>Bin</th>\n",
       "      <th>classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2</td>\n",
       "      <td>4Wh2oV_72h_scaffold_529029</td>\n",
       "      <td>96.970</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>3511</td>\n",
       "      <td>3479</td>\n",
       "      <td>33</td>\n",
       "      <td>3.140000e-07</td>\n",
       "      <td>56.5</td>\n",
       "      <td>H1-18-all-fractions_metab_25.Bacteria_70_9</td>\n",
       "      <td>H1-18-all-fractions_k255_4602126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2</td>\n",
       "      <td>4Wh2oV_72h_scaffold_695820</td>\n",
       "      <td>96.970</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>594</td>\n",
       "      <td>626</td>\n",
       "      <td>33</td>\n",
       "      <td>3.140000e-07</td>\n",
       "      <td>56.5</td>\n",
       "      <td>H1-18-all-fractions_metab_25.Bacteria_70_9</td>\n",
       "      <td>H1-18-all-fractions_k255_4602126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2</td>\n",
       "      <td>4Wh2oV_48h_scaffold_936400</td>\n",
       "      <td>96.970</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>2686</td>\n",
       "      <td>2654</td>\n",
       "      <td>33</td>\n",
       "      <td>3.140000e-07</td>\n",
       "      <td>56.5</td>\n",
       "      <td>H1-18-all-fractions_metab_25.Bacteria_70_9</td>\n",
       "      <td>H1-18-all-fractions_k255_4602126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2</td>\n",
       "      <td>4Wh2oV_0h_scaffold_637225</td>\n",
       "      <td>96.970</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>975</td>\n",
       "      <td>1007</td>\n",
       "      <td>33</td>\n",
       "      <td>3.140000e-07</td>\n",
       "      <td>56.5</td>\n",
       "      <td>H1-18-all-fractions_metab_25.Bacteria_70_9</td>\n",
       "      <td>H1-18-all-fractions_k255_4602126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2</td>\n",
       "      <td>4Wh2oV_48h_scaffold_716693</td>\n",
       "      <td>100.000</td>\n",
       "      <td>85</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>1633</td>\n",
       "      <td>1606</td>\n",
       "      <td>28</td>\n",
       "      <td>4.060000e-06</td>\n",
       "      <td>52.8</td>\n",
       "      <td>H1-18-all-fractions_metab_25.Bacteria_70_9</td>\n",
       "      <td>H1-18-all-fractions_k255_4602126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>H1-18-all-fractions_k255_4902837_CRISPR_1_spacer_4</td>\n",
       "      <td>4Wh2oV_3h_scaffold_1843257</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>108</td>\n",
       "      <td>139</td>\n",
       "      <td>32</td>\n",
       "      <td>2.120000e-08</td>\n",
       "      <td>60.2</td>\n",
       "      <td>H1-18-all-fractions_metab_63.Actinobacteria_71_13_spacers.fa</td>\n",
       "      <td>H1-18-all-fractions_k255_4902837</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H3-18-all-fractions_k255_4574136_CRISPR_1_spacer_1</td>\n",
       "      <td>4Wh2oV_24h_scaffold_1787552</td>\n",
       "      <td>95.122</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "      <td>41</td>\n",
       "      <td>89</td>\n",
       "      <td>129</td>\n",
       "      <td>41</td>\n",
       "      <td>1.110000e-09</td>\n",
       "      <td>65.8</td>\n",
       "      <td>H3-18-all-fractions_conc_96.Burkholderiales_71_8_spacers.fa</td>\n",
       "      <td>H3-18-all-fractions_k255_4574136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_1</td>\n",
       "      <td>4Wh2oV_24h_scaffold_1626649</td>\n",
       "      <td>100.000</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>37</td>\n",
       "      <td>1092</td>\n",
       "      <td>1121</td>\n",
       "      <td>30</td>\n",
       "      <td>4.710000e-07</td>\n",
       "      <td>56.5</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_2</td>\n",
       "      <td>4Wh2oV_3h_scaffold_1562313</td>\n",
       "      <td>100.000</td>\n",
       "      <td>80</td>\n",
       "      <td>8</td>\n",
       "      <td>35</td>\n",
       "      <td>834</td>\n",
       "      <td>861</td>\n",
       "      <td>28</td>\n",
       "      <td>5.080000e-06</td>\n",
       "      <td>52.8</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_2</td>\n",
       "      <td>4Wh2oV_24h_scaffold_960306</td>\n",
       "      <td>100.000</td>\n",
       "      <td>80</td>\n",
       "      <td>8</td>\n",
       "      <td>35</td>\n",
       "      <td>687</td>\n",
       "      <td>714</td>\n",
       "      <td>28</td>\n",
       "      <td>5.080000e-06</td>\n",
       "      <td>52.8</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_3</td>\n",
       "      <td>4Wh2oV_3h_scaffold_1894264</td>\n",
       "      <td>100.000</td>\n",
       "      <td>85</td>\n",
       "      <td>6</td>\n",
       "      <td>33</td>\n",
       "      <td>843</td>\n",
       "      <td>870</td>\n",
       "      <td>28</td>\n",
       "      <td>4.060000e-06</td>\n",
       "      <td>52.8</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_3</td>\n",
       "      <td>4Wh2oV_24h_scaffold_467538</td>\n",
       "      <td>100.000</td>\n",
       "      <td>85</td>\n",
       "      <td>6</td>\n",
       "      <td>33</td>\n",
       "      <td>556</td>\n",
       "      <td>529</td>\n",
       "      <td>28</td>\n",
       "      <td>4.060000e-06</td>\n",
       "      <td>52.8</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_3</td>\n",
       "      <td>4Wh2oV_0h_scaffold_1703601</td>\n",
       "      <td>100.000</td>\n",
       "      <td>85</td>\n",
       "      <td>6</td>\n",
       "      <td>33</td>\n",
       "      <td>2052</td>\n",
       "      <td>2025</td>\n",
       "      <td>28</td>\n",
       "      <td>4.060000e-06</td>\n",
       "      <td>52.8</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_8</td>\n",
       "      <td>4Wh2oV_24h_scaffold_1255333</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>1404</td>\n",
       "      <td>1436</td>\n",
       "      <td>33</td>\n",
       "      <td>6.750000e-09</td>\n",
       "      <td>62.1</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_8</td>\n",
       "      <td>4Wh2oV_3h_scaffold_356956</td>\n",
       "      <td>96.970</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>2926</td>\n",
       "      <td>2894</td>\n",
       "      <td>33</td>\n",
       "      <td>3.140000e-07</td>\n",
       "      <td>56.5</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_8</td>\n",
       "      <td>4Wh2oV_24h_scaffold_1500359</td>\n",
       "      <td>96.970</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>3490</td>\n",
       "      <td>3458</td>\n",
       "      <td>33</td>\n",
       "      <td>3.140000e-07</td>\n",
       "      <td>56.5</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_9</td>\n",
       "      <td>4Wh2oV_24h_scaffold_221362</td>\n",
       "      <td>100.000</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>920</td>\n",
       "      <td>886</td>\n",
       "      <td>35</td>\n",
       "      <td>6.520000e-10</td>\n",
       "      <td>65.8</td>\n",
       "      <td>S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa</td>\n",
       "      <td>S3-16-all-fractions_k255_1197418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                qseqid  \\\n",
       "0   H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2   \n",
       "1   H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2   \n",
       "2   H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2   \n",
       "3   H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2   \n",
       "4   H1-18-all-fractions_k255_4602126_CRISPR_1_spacer_2   \n",
       "5   H1-18-all-fractions_k255_4902837_CRISPR_1_spacer_4   \n",
       "6   H3-18-all-fractions_k255_4574136_CRISPR_1_spacer_1   \n",
       "7   S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_1   \n",
       "8   S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_2   \n",
       "9   S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_2   \n",
       "10  S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_3   \n",
       "11  S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_3   \n",
       "12  S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_3   \n",
       "13  S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_8   \n",
       "14  S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_8   \n",
       "15  S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_8   \n",
       "16  S3-16-all-fractions_k255_1197418_CRISPR_1_spacer_9   \n",
       "\n",
       "                         sseqid   pident  qcovs  qstart  qend  sstart  send  \\\n",
       "0    4Wh2oV_72h_scaffold_529029   96.970    100       1    33    3511  3479   \n",
       "1    4Wh2oV_72h_scaffold_695820   96.970    100       1    33     594   626   \n",
       "2    4Wh2oV_48h_scaffold_936400   96.970    100       1    33    2686  2654   \n",
       "3     4Wh2oV_0h_scaffold_637225   96.970    100       1    33     975  1007   \n",
       "4    4Wh2oV_48h_scaffold_716693  100.000     85       1    28    1633  1606   \n",
       "5    4Wh2oV_3h_scaffold_1843257  100.000    100       1    32     108   139   \n",
       "6   4Wh2oV_24h_scaffold_1787552   95.122     95       1    41      89   129   \n",
       "7   4Wh2oV_24h_scaffold_1626649  100.000     81       8    37    1092  1121   \n",
       "8    4Wh2oV_3h_scaffold_1562313  100.000     80       8    35     834   861   \n",
       "9    4Wh2oV_24h_scaffold_960306  100.000     80       8    35     687   714   \n",
       "10   4Wh2oV_3h_scaffold_1894264  100.000     85       6    33     843   870   \n",
       "11   4Wh2oV_24h_scaffold_467538  100.000     85       6    33     556   529   \n",
       "12   4Wh2oV_0h_scaffold_1703601  100.000     85       6    33    2052  2025   \n",
       "13  4Wh2oV_24h_scaffold_1255333  100.000    100       1    33    1404  1436   \n",
       "14    4Wh2oV_3h_scaffold_356956   96.970    100       1    33    2926  2894   \n",
       "15  4Wh2oV_24h_scaffold_1500359   96.970    100       1    33    3490  3458   \n",
       "16   4Wh2oV_24h_scaffold_221362  100.000    100       1    35     920   886   \n",
       "\n",
       "    length        evalue  bitscore  \\\n",
       "0       33  3.140000e-07      56.5   \n",
       "1       33  3.140000e-07      56.5   \n",
       "2       33  3.140000e-07      56.5   \n",
       "3       33  3.140000e-07      56.5   \n",
       "4       28  4.060000e-06      52.8   \n",
       "5       32  2.120000e-08      60.2   \n",
       "6       41  1.110000e-09      65.8   \n",
       "7       30  4.710000e-07      56.5   \n",
       "8       28  5.080000e-06      52.8   \n",
       "9       28  5.080000e-06      52.8   \n",
       "10      28  4.060000e-06      52.8   \n",
       "11      28  4.060000e-06      52.8   \n",
       "12      28  4.060000e-06      52.8   \n",
       "13      33  6.750000e-09      62.1   \n",
       "14      33  3.140000e-07      56.5   \n",
       "15      33  3.140000e-07      56.5   \n",
       "16      35  6.520000e-10      65.8   \n",
       "\n",
       "                                                               MAG  \\\n",
       "0                       H1-18-all-fractions_metab_25.Bacteria_70_9   \n",
       "1                       H1-18-all-fractions_metab_25.Bacteria_70_9   \n",
       "2                       H1-18-all-fractions_metab_25.Bacteria_70_9   \n",
       "3                       H1-18-all-fractions_metab_25.Bacteria_70_9   \n",
       "4                       H1-18-all-fractions_metab_25.Bacteria_70_9   \n",
       "5     H1-18-all-fractions_metab_63.Actinobacteria_71_13_spacers.fa   \n",
       "6      H3-18-all-fractions_conc_96.Burkholderiales_71_8_spacers.fa   \n",
       "7   S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "8   S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "9   S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "10  S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "11  S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "12  S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "13  S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "14  S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "15  S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "16  S3-16-all-fractions_metab_conc_21.Rhizobiales_66_23_spacers.fa   \n",
       "\n",
       "                         scaffold_id scaffold  Bin classification  \n",
       "0   H1-18-all-fractions_k255_4602126      NaN  NaN            NaN  \n",
       "1   H1-18-all-fractions_k255_4602126      NaN  NaN            NaN  \n",
       "2   H1-18-all-fractions_k255_4602126      NaN  NaN            NaN  \n",
       "3   H1-18-all-fractions_k255_4602126      NaN  NaN            NaN  \n",
       "4   H1-18-all-fractions_k255_4602126      NaN  NaN            NaN  \n",
       "5   H1-18-all-fractions_k255_4902837      NaN  NaN            NaN  \n",
       "6   H3-18-all-fractions_k255_4574136      NaN  NaN            NaN  \n",
       "7   S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  \n",
       "8   S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  \n",
       "9   S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  \n",
       "10  S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  \n",
       "11  S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  \n",
       "12  S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  \n",
       "13  S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  \n",
       "14  S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  \n",
       "15  S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  \n",
       "16  S3-16-all-fractions_k255_1197418      NaN  NaN            NaN  "
      ]
     },
     "execution_count": 410,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wsip_mag.merge(s2b,on='scaffold_id',how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "wsip_mag['votu'] = wsip_mag['sseqid']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 680,
   "metadata": {},
   "outputs": [],
   "source": [
    "wsip_contig_file = '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/wSIP_contigs_spacers_vs_4Wh2oV_blast_db.tsv'\n",
    "wsip_contig = pd.read_csv(wsip_contig_file,sep='\\t')\n",
    "wsip_contig['scaffold_id'] = wsip_contig['qseqid'].str.split(\"_C\").str[0]\n",
    "wsip_contig['array'] = wsip_contig['qseqid'].str.split(\"CRISPR\").str[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 681,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge kaiju taxa prediction with wsip contigs that have crispr matches to vOTUs\n",
    "wsip_contig = pd.merge(wsip_contig, kaiju_out, left_on='scaffold_id',right_on='kaiju_scaffold',how='left')\n",
    "# kaiju_out[kaiju_out['kaiju_scaffold'].isin(wsip_contig['scaffold_id'].unique())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 682,
   "metadata": {},
   "outputs": [],
   "source": [
    "wsip_contig['votu'] = wsip_contig['sseqid']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 684,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pid</th>\n",
       "      <th>qcov</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>length</th>\n",
       "      <th>e-value</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>scaffold_id</th>\n",
       "      <th>array</th>\n",
       "      <th>classified</th>\n",
       "      <th>kaiju_scaffold</th>\n",
       "      <th>ncbi_taxid</th>\n",
       "      <th>kaiju_taxonomy</th>\n",
       "      <th>votu</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>H3-18-all-fractions_k255_4574136_CRISPR_17_spacer_1</td>\n",
       "      <td>4Wh2oV_24h_scaffold_1787552</td>\n",
       "      <td>95.122</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "      <td>41</td>\n",
       "      <td>89</td>\n",
       "      <td>129</td>\n",
       "      <td>41</td>\n",
       "      <td>1.110000e-09</td>\n",
       "      <td>65.8</td>\n",
       "      <td>H3-18-all-fractions_k255_4574136</td>\n",
       "      <td>_17_spacer_1</td>\n",
       "      <td>C</td>\n",
       "      <td>H3-18-all-fractions_k255_4574136</td>\n",
       "      <td>983917</td>\n",
       "      <td>cellular organisms; Bacteria; Proteobacteria; Betaproteobacteria; Burkholderiales; Burkholderiales genera incertae sedis; Rubrivivax; Rubrivivax gelatinosus</td>\n",
       "      <td>4Wh2oV_24h_scaffold_1787552</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 qseqid  \\\n",
       "68  H3-18-all-fractions_k255_4574136_CRISPR_17_spacer_1   \n",
       "\n",
       "                         sseqid     pid  qcov  qstart  qend  sstart  send  \\\n",
       "68  4Wh2oV_24h_scaffold_1787552  95.122    95       1    41      89   129   \n",
       "\n",
       "    length       e-value  bitscore                       scaffold_id  \\\n",
       "68      41  1.110000e-09      65.8  H3-18-all-fractions_k255_4574136   \n",
       "\n",
       "           array classified                    kaiju_scaffold  ncbi_taxid  \\\n",
       "68  _17_spacer_1          C  H3-18-all-fractions_k255_4574136      983917   \n",
       "\n",
       "                                                                                                                                                  kaiju_taxonomy  \\\n",
       "68  cellular organisms; Bacteria; Proteobacteria; Betaproteobacteria; Burkholderiales; Burkholderiales genera incertae sedis; Rubrivivax; Rubrivivax gelatinosus   \n",
       "\n",
       "                           votu  \n",
       "68  4Wh2oV_24h_scaffold_1787552  "
      ]
     },
     "execution_count": 684,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wsip_contig[wsip_contig['scaffold_id'].isin(wsip_mag['scaffold_id'].unique())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1493,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "vcontact = pd.read_csv(\"/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/viral_id/vcontact_out/merged_df.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Combining all host predictions\n",
    "Generate df with all spacer matches and taxonomy in the format: \n",
    "\n",
    "vOTU / spacer / spacer taxonomy [strain / taxon info]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Combine all unbinned contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 727,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['qseqid', 'sseqid', 'pident', 'qcovs', 'qstart', 'qend', 'sstart',\n",
       "       'send', 'length', 'evalue', 'bitscore', 'scaffold_id', 'array',\n",
       "       'classified', 'kaiju_scaffold', 'ncbi_taxid', 'kaiju_taxonomy', 'votu',\n",
       "       'source'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 727,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# setting unbinned contigs\n",
    "unbinned_hosts = mg_spacers_blastout_taxa.rename(columns={'mg_contig':'scaffold_id','e-value':'evalue',\n",
    "                                                    'pid':'pident','qcov':'qcovs'})\n",
    "unbinned_hosts['source'] = 'unbinned'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NOTES:\n",
    "* don't rename taxonomy yet, because for some they are viral and need to get host\n",
    "* duplicates in mg_spacers blastout are from multiple spacer hits to contig"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## combine blast results of MAGs i.e. using scaff2bin file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_cols = ['qseqid', 'sseqid', 'pident', 'qcovs', 'qstart', 'qend', 'sstart',\n",
    "       'send', 'length', 'evalue', 'bitscore']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [],
   "source": [
    "mag_crispr_blast = pd.concat(mag_spacer_hits,vOTU_spacer_MAGs])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# s2bin file is a scaffold to bin file i.e. for each contig in a MAG it's host taxonomy\n",
    "# merge scaffold to bin file to establish vOTU taxonomy based on MAG taxonomy\n",
    "mag_crispr_blast = mag_crispr_blast.merge(s2b[['Bin','taxa']]\\\n",
    "                                              .rename(columns={'taxa':'rep_taxa'})\\\n",
    "                                              .drop_duplicates(),\n",
    "                                        how='left',\n",
    "                                        validate='m:1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 622,
   "metadata": {},
   "outputs": [],
   "source": [
    "mag_crispr_blast['source'] = 'MAG'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Format spacer db - vOTU matches to align with other host match strategies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 797,
   "metadata": {},
   "outputs": [],
   "source": [
    "spacer_db_blast_result['votu'] = spacer_db_blast_result['sseqid'] \n",
    "spacer_db_blast_result['source'] = 'spacerDB'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 626,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_cols = [\n",
    "'votu','taxa', 'source','qseqid', 'sseqid', 'pident', 'qcovs', 'qstart', 'qend', 'sstart',\n",
    "'send', 'length', 'evalue', 'bitscore',\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Format vOTU taxonomy based on kaiju output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 823,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-823-120eabb5eab6>:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  kaiju_votus_taxa['source'] = 'kaiju'\n",
      "<ipython-input-823-120eabb5eab6>:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  kaiju_votus_taxa['votu'] = kaiju_votus_taxa['kaiju_scaffold']\n"
     ]
    }
   ],
   "source": [
    "kaiju_votus_taxa = kaiju_out[kaiju_out['kaiju_scaffold'].isin(kaiju_vOTUs['kaiju_scaffold'].unique())]\n",
    "kaiju_votus_taxa['source'] = 'kaiju'\n",
    "kaiju_votus_taxa['votu'] = kaiju_votus_taxa['kaiju_scaffold']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 734,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# combined df of all types of crispr hits\n",
    "all_crispr_hits_long = pd.concat([\n",
    "    mag_crispr_blast[common_cols],\n",
    "    unbinned_hosts.rename(columns={'kaiju_taxonomy':'taxa'})[common_cols],\n",
    "    spacer_db_blast_result.rename(columns={\"taxonomy\":'taxa'})[['votu','taxa', 'source','qseqid', 'sseqid', 'pident', 'qcovs', 'qstart', 'qend', 'sstart',\n",
    "       'send', 'length', 'evalue', 'bitscore',\n",
    "                                                               ]]\n",
    "]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Gather host lineage from virushost db for unbinned contig spacers / votus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: \n",
    "* many viruses have multiple hosts – this merge must be done on virus name to be more specific (rather than host)\n",
    "* This is required for when Kaiju marks vOTU taxonomy as being viral i.e., vOTU has a reference genome documented in NCBI as viral – this is the case for a subset of total vOTUs identified (speaks to viral diversity captured here compared to reference db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 755,
   "metadata": {},
   "outputs": [],
   "source": [
    "kaiju_votus_taxa_nonull = kaiju_votus_taxa[kaiju_votus_taxa['taxa'].notna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "most conservative method is maintain duplicate taxonomies per vOTU by source and allow the groupby function to determine consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 771,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge virushostdb by changing the taxonomy column to virus name (more specific than lineage), \n",
    "# use combine_first to get either taxa (should be bacterial) or host lineage as taxa\n",
    "kaiju_votus_taxa_host = kaiju_votus_taxa_nonull.rename(columns={'kaiju_taxonomy':'virus name'})\\\n",
    "    .merge(virushostdb[['virus name','host lineage']],\n",
    "                     on='virus name',\n",
    "                     how='left')\n",
    "# rename taxa to draft column because will use combine first to get new taxa column\n",
    "kaiju_votus_taxa_host.rename(columns={'taxa':'taxa_draft'},inplace=True)\n",
    "# use combine first to get distinct taxa column\n",
    "kaiju_votus_taxa_host['taxa'] = kaiju_votus_taxa_host[['host lineage']].rename(\n",
    "    columns={'host lineage':'taxa'}).combine_first(\n",
    "    kaiju_votus_taxa_host[['taxa_draft']].rename(\n",
    "    columns={'taxa_draft':'taxa'}))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make [massive] df: all types of taxonomy calls per vOTU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 788,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter crispr spacer hits to qcovs > 90%\n",
    "all_crispr_hits_long_filter = all_crispr_hits_long[all_crispr_hits_long['qcovs']>=90]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 862,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add kaiju votu taxas (w host too!) to all spacer db hits \n",
    "all_votu_taxonomy = pd.concat([all_crispr_hits_long_filter[['votu','taxa','source']],\n",
    "                                       kaiju_votus_taxa_host[['votu','taxa','source']]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 864,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop duplicates - duplicates derive from multiple spacer hits with same taxonomy\n",
    "all_votu_taxonomy_filter_simplified = all_votu_taxonomy.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 865,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>votu</th>\n",
       "      <th>taxa</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4Wh2oV_72h_scaffold_529029</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>MAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4Wh2oV_72h_scaffold_695820</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>MAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4Wh2oV_48h_scaffold_936400</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>MAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         votu      taxa source\n",
       "0  4Wh2oV_72h_scaffold_529029  Bacteria    MAG\n",
       "1  4Wh2oV_72h_scaffold_695820  Bacteria    MAG\n",
       "2  4Wh2oV_48h_scaffold_936400  Bacteria    MAG"
      ]
     },
     "execution_count": 865,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_votu_taxonomy_filter_simplified.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example case vOTU with multiple taxonomy calls from multiple sources \n",
    "4Wh2oV_0h_scaffold_1602168 (https://ggkbase.berkeley.edu/organisms/585617)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 867,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>votu</th>\n",
       "      <th>taxa</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>4Wh2oV_0h_scaffold_1602168</td>\n",
       "      <td>['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Klebsiella']</td>\n",
       "      <td>spacerDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>4Wh2oV_0h_scaffold_1602168</td>\n",
       "      <td>['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']</td>\n",
       "      <td>spacerDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>4Wh2oV_0h_scaffold_1602168</td>\n",
       "      <td>['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Methylococcales', 'Methylococcaceae', 'Methylomicrobium']</td>\n",
       "      <td>spacerDB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20336</th>\n",
       "      <td>4Wh2oV_0h_scaffold_1602168</td>\n",
       "      <td>cellular organisms; Bacteria; Proteobacteria; Alphaproteobacteria; Hyphomicrobiales; Rhizobiaceae; Sinorhizobium/Ensifer group; Sinorhizobium; Sinorhizobium meliloti</td>\n",
       "      <td>kaiju</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             votu  \\\n",
       "101    4Wh2oV_0h_scaffold_1602168   \n",
       "169    4Wh2oV_0h_scaffold_1602168   \n",
       "993    4Wh2oV_0h_scaffold_1602168   \n",
       "20336  4Wh2oV_0h_scaffold_1602168   \n",
       "\n",
       "                                                                                                                                                                        taxa  \\\n",
       "101                                                            ['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Enterobacterales', 'Enterobacteriaceae', 'Klebsiella']   \n",
       "169                                                              ['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Pseudomonadales', 'Pseudomonadaceae', 'Pseudomonas']   \n",
       "993                                                         ['Bacteria', 'Proteobacteria', 'Gammaproteobacteria', 'Methylococcales', 'Methylococcaceae', 'Methylomicrobium']   \n",
       "20336  cellular organisms; Bacteria; Proteobacteria; Alphaproteobacteria; Hyphomicrobiales; Rhizobiaceae; Sinorhizobium/Ensifer group; Sinorhizobium; Sinorhizobium meliloti   \n",
       "\n",
       "         source  \n",
       "101    spacerDB  \n",
       "169    spacerDB  \n",
       "993    spacerDB  \n",
       "20336     kaiju  "
      ]
     },
     "execution_count": 867,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_votu_taxonomy_filter_simplified[all_votu_taxonomy_filter_simplified['votu']=='4Wh2oV_0h_scaffold_1602168']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Groupby function that establishes consensus taxonomy per vOTU from multiple sources/taxonomy calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1073,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def least_common_string(list_of_inputs):\n",
    "    '''read in a column of taxons as a sorted (by length) – referred to here as list of inputs \n",
    "    clean list of lists\n",
    "    for each list in list of lists iterate through  \n",
    "    update with intersecting elements\n",
    "    will output a list of consensus of lists, order is preserved \n",
    "    use with groupby function on taxa column \n",
    "    where data is formatted: seq_name; taxonomy (as list); source'''\n",
    "    \n",
    "    # turn every string into list and return cleaned list \n",
    "    cleaned_inputs = []\n",
    "    for item in list_of_inputs:\n",
    "        if \";\" in item:\n",
    "            delim=\";\"\n",
    "        else: delim=','\n",
    "        cleaned_list = item.replace(\"[\",\"\").replace(\"]\",\"\").replace(\"'\",\"\").split(delim)\n",
    "        cleaned_inputs.append([x.strip(\" \") for x in cleaned_list])\n",
    "    # return cleaned_inputs\n",
    "\n",
    "    if len(cleaned_inputs) >1:\n",
    "        # within each list (cleaned_inputs) find shortest list to use as reference\n",
    "        # list of lists sorted by the length of each item \n",
    "        sorted_cleaned_inputs = sorted(cleaned_inputs, key=lambda x: len(x))\n",
    "        reference = list(sorted_cleaned_inputs[0])\n",
    "\n",
    "        # use the set function – intersection_update \n",
    "        # iterate through list (ordered by length) \n",
    "        # update to common set of list items as iterate through\n",
    "        for item in sorted_cleaned_inputs[1:]:\n",
    "            reference = [x for x in reference if x in item]\n",
    "            \n",
    "        return \",\".join(list(reference))\n",
    "    else: return \",\".join(cleaned_inputs[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 812,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Bacteria,Proteobacteria'"
      ]
     },
     "execution_count": 812,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "least_common_string(\n",
    "    all_votu_taxonomy_filter_simplified[\n",
    "        all_votu_taxonomy_filter_simplified['votu']=='4Wh2oV_0h_scaffold_1602168']['taxa']\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "manual inspection of vOTU shows sequence homology to Alphaproteobacteria, interestingly CRISPR spacers match different orders of Gammaproteobacteria, consensus taxonomy is thus most conservative though lacking greater taxonomic resolution "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Establish consensus host taxonomies for vOTUs\n",
    "Run least_common_strings function on the \"massive\" df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 868,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_votu_taxonomy_filter_simplified.to_csv(    '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/host_info/all_vOTU_host_predictions_xsource2.csv',\n",
    "index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 871,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run function and create new column – beware it's just for not null taxa calls \n",
    "consensus_taxonomy_votus = all_votu_taxonomy_filter_simplified[\n",
    "    ~all_votu_taxonomy_filter_simplified['taxa'].isna()\n",
    "                                      ].groupby(\n",
    "    'votu',as_index=False).aggregate(\n",
    "    {'taxa':least_common_string})\n",
    "consensus_taxonomy_votus['taxa'] = consensus_taxonomy_votus['taxa'].str.replace(\n",
    "                                            'cellular organisms,','').str.strip()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "some funky edge cases... \n",
    "###### Fix ncbi taxid 979 cellulophaga lytica\n",
    "https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=979"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 877,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>votu</th>\n",
       "      <th>taxa</th>\n",
       "      <th>consensus_domain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>4Wh2oV_0h_scaffold_1090805</td>\n",
       "      <td>Cellulophaga lytica</td>\n",
       "      <td>Cellulophaga lytica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18928</th>\n",
       "      <td>4Wh2oV_3h_scaffold_902073</td>\n",
       "      <td>Cellulophaga lytica</td>\n",
       "      <td>Cellulophaga lytica</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             votu                 taxa     consensus_domain\n",
       "544    4Wh2oV_0h_scaffold_1090805  Cellulophaga lytica  Cellulophaga lytica\n",
       "18928   4Wh2oV_3h_scaffold_902073  Cellulophaga lytica  Cellulophaga lytica"
      ]
     },
     "execution_count": 877,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_taxonomy_votus[consensus_taxonomy_votus['consensus_domain']=='Cellulophaga lytica']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 878,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# fix cellulophaga lytica – ncbi taxid 979: \n",
    "# Bacteria; FCB group; Bacteroidetes/Chlorobi group; Bacteroidetes; Flavobacteriia; Flavobacteriales; Flavobacteriaceae; Cellulophaga\n",
    "consensus_taxonomy_votus['taxa'] = consensus_taxonomy_votus['taxa'].str.replace(\n",
    "    \"Cellulophaga lytica\",\n",
    "\"Bacteria,FCB group,Bacteroidetes/Chlorobi group,Bacteroidetes,Flavobacteriia,Flavobacteriales,Flavobacteriaceae,Cellulophaga lytica\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "###### Run consensus domain post fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 885,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_taxonomy_votus['consensus_domain'] = consensus_taxonomy_votus[\n",
    "                                    'taxa'].str.split(',').str[0].str.replace(\n",
    "    r'^\\s*$','unknown',regex=True\n",
    ").str.replace(\n",
    "    \"Viruses\",\"unknown\"\n",
    ").str.replace(\n",
    "    \"cellular organisms\",\"unknown\").fillna('unknown')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 886,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Bacteria', 'Archaea', 'unknown', 'Eukaryota'], dtype=object)"
      ]
     },
     "execution_count": 886,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_taxonomy_votus['consensus_domain'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 887,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>consensus_domain</th>\n",
       "      <th>votu</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>24248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>unknown</td>\n",
       "      <td>440</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  consensus_domain   votu\n",
       "0          Archaea    212\n",
       "1         Bacteria  24248\n",
       "2        Eukaryota      3\n",
       "3          unknown    440"
      ]
     },
     "execution_count": 887,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_taxonomy_votus.groupby('consensus_domain',as_index=False).agg({'votu':'count'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Some \"viruses\" still didn't receive a host name from hostvirusdb i.e. votus picked up by kaiju, can manually fix because not that many..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 892,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manual look up to create dictionary for host of viral lineages \n",
    "manual_host_dict = {\n",
    "'Viruses,Duplodnaviria,Heunggongvirae,Uroviricota,Caudoviricetes,Caudovirales,Siphoviridae,Timquatrovirus':\n",
    "    'Bacteria,Terrabacteria group,Actinobacteria,Actinomycetia,Corynebacteriales,Mycobacteriaceae',\n",
    "'Viruses,Duplodnaviria,Heunggongvirae,Uroviricota,Caudoviricetes,Caudovirales,Siphoviridae,Cheoctovirus':\n",
    "    'Bacteria,Terrabacteria group,Actinobacteria,Actinomycetia,Corynebacteriales,Mycobacteriaceae',\n",
    "'Viruses,Duplodnaviria,Heunggongvirae,Uroviricota,Caudoviricetes,Caudovirales,Siphoviridae,Avanivirus':\n",
    "    'Bacteria,Terrabacteria group,Actinobacteria,Actinomycetia,Corynebacteriales,Mycobacteriaceae',\n",
    "'Viruses,Duplodnaviria,Heunggongvirae,Uroviricota,Caudoviricetes,Caudovirales,Siphoviridae,Bronvirus':\n",
    "    'Bacteria,Terrabacteria group,Actinobacteria,Actinomycetia,Corynebacteriales,Mycobacteriaceae',\n",
    "'Viruses,Duplodnaviria,Heunggongvirae,Uroviricota,Caudoviricetes,Caudovirales,Siphoviridae,Fromanvirus':\n",
    "    'Bacteria,Terrabacteria group,Actinobacteria,Actinomycetia,Corynebacteriales,Mycobacteriaceae'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 897,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace values \n",
    "consensus_taxonomy_votus = consensus_taxonomy_votus.replace({'taxa':manual_host_dict})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 905,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_taxonomy_votus['consensus_domain'] = consensus_taxonomy_votus[\n",
    "                                    'taxa'].str.split(',').str[0].str.replace(\n",
    "    r'^\\s*$','unknown',regex=True\n",
    ").str.replace(\n",
    "    \"Viruses\",\"unknown\"\n",
    ").str.replace(\n",
    "    \"cellular organisms\",\"unknown\").fillna('unknown')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 907,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>consensus_domain</th>\n",
       "      <th>votu</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>24329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>unknown</td>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  consensus_domain   votu\n",
       "0          Archaea    212\n",
       "1         Bacteria  24329\n",
       "2        Eukaryota      3\n",
       "3          unknown    359"
      ]
     },
     "execution_count": 907,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_taxonomy_votus.groupby('consensus_domain',as_index=False).agg({'votu':'count'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### write out ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 909,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_taxonomy_votus = pd.read_csv(\n",
    "    '/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/votu_consensus_taxa2.tsv',\n",
    "                                       sep='\\t').fillna('unknown')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 913,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEwCAYAAABYJG2OAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAZbElEQVR4nO3de5RlZX3m8e/T7QUVBSNiUESIrShGFG0JLi/xlgwMthgVATWIw4CoeIuimHHNaNSMSjRRNJiOEtAgjuANFMTgBRRQaBS52FEJhgijNhdl4UjQxt/8sXdtDm11d3V17d61q76ftXr12e85tftXZ1XXc/b7vvt9U1VIkgSwZOgCJEnzh6EgSeoYCpKkjqEgSeoYCpKkjqEgSercaegCNsd2221XO++889BlSNKoXHzxxddX1X2ne27UobDzzjuzatWqocuQpFFJcvX6nrP7SJLUMRQkSR1DQZLUGWUoJFmRZOVNN900dCmStKCMMhSq6vSqOnybbbYZuhRJWlBGGQqSpH4YCpKkjqEgSeqM+uY1bVlrjnvD0CXMG9u/7N1DlyD1wisFSVLHUJAkdQwFSVJnlKHgzWuS1I9RhoI3r0lSP0YZCpKkfhgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqTOvAqFJPdIsirJM4euRZIWo15DIcnxSdYkuXyd9r2TfD/JlUmOnnjqjcAn+6xJkrR+fV8pnADsPdmQZCnwQWAfYDfgoCS7JfkT4HvAmp5rkiStR6/bcVbVuUl2Xqd5T+DKqroKIMkngP2ArYF70ATFLUnOqKrfrnvOJIcDhwPstNNO/RUvSYvQEHs0PwD48cTxNcAfVdWRAEkOAa6fLhAAqmolsBJg+fLl1W+pkrS4DBEKG1RVJwxdgyQtVkPMProWeODE8Y5t24y585ok9WOIULgIeEiSXZLcBTgQOG1TTuDOa5LUj76npJ4MXADsmuSaJIdW1VrgSOAsYDXwyaq6YhPP65WCJPWg79lHB62n/QzgjM047+nA6cuXLz9stueQJP2ueXVHsyRpWKMMBbuPJKkfowwFB5olqR+jDAVJUj8MBUlSZ5Sh4JiCJPVjlKHgmIIk9WOUoSBJ6oehIEnqjDIUHFOQpH6MMhQcU5CkfowyFCRJ/TAUJEkdQ0GS1BllKDjQLEn9GGUoONAsSf0YZShIkvphKEiSOoaCJKljKEiSOqMMBWcfSVI/RhkKzj6SpH6MMhQkSf0wFCRJHUNBktQxFCRJHUNBktQxFCRJHUNBktQZZSh485ok9WOUoeDNa5LUj1GGgiSpH4aCJKljKEiSOoaCJKljKEiSOoaCJKljKEiSOoaCJKljKEiSOvMmFJI8PMmHkpya5GVD1yNJi1GvoZDk+CRrkly+TvveSb6f5MokRwNU1eqqOgJ4PvCEPuuSJE2v7yuFE4C9JxuSLAU+COwD7AYclGS39rlnAV8Azui5LknSNHoNhao6F7hxneY9gSur6qqq+jXwCWC/9vWnVdU+wAv7rEuSNL07DfBvPgD48cTxNcAfJXkK8BzgrmzgSiHJ4cDhADvttFN/VUrSIjREKEyrqr4GfG0Gr1sJrARYvnx59VuVJC0uQ8w+uhZ44MTxjm3bjLnJjiT1Y4hQuAh4SJJdktwFOBA4bVNO4CY7ktSPvqekngxcAOya5Jokh1bVWuBI4CxgNfDJqrqizzokSTPT65hCVR20nvYz2Ixpp0lWACuWLVs221NIkqYxb+5o3hR2H0lSP0YZCpKkfowyFJx9JEn9GGUo2H0kSf0YZShIkvoxylCw+0iS+jHKULD7SJL6McpQkCT1w1CQJHVGGQqOKUhSP0YZCo4pSFI/RhkKkqR+GAqSpI6hIEnqbHTp7CRPnq69qs6d+3JmxqWzJakfM9lP4aiJx1sBewIXA0/rpaIZqKrTgdOXL19+2FA1SNJCtNFQqKoVk8dJHgj8XW8VSZIGM5ud164BHj7XhUjS5rjyuHOGLmFeWPayP96sr5/JmMKxQLWHS4BHA9/erH9VkjQvzeRKYdXE47XAyVV1Xk/1SJIGNJNQ2Laq3jfZkOTV67ZtSc4+kqR+zOQ+hRdP03bIHNexSVzmQpL6sd4rhSQHAS8Adkly2sRT9wJu7LswSdKWt6Huo/OBnwDbAe+ZaL8ZuLTPoiRJw1hvKFTV1cDVwOOT3A94XPvU6qpauyWKkyRtWRsdU0iyP3AhsD/wfOBbSZ7Xd2GSpC1vJrOP3gw8rqrWACS5L3A2cGqfhUmStryZzD5aMhUIrRtm+HWSpJGZyZXCmUnOAk5ujw8AzuivpLnzujM/OnQJ88Z79jl46BIkjcBMQmEN8M80y1sArKyqz/RX0sZ585ok9WMm3UD3AI6mWTL7RzRTVQflzWuS1I+NhkJVvbWqHgG8AtgBOCfJ2b1XJkna4jZlwHgN8FOagebt+ylHkjSkmdyn8PIkXwO+DNwHOKyqdu+7MEnSljeTgeYHAq+pqkv6LkaSNKyZbMf5pi1RiCRpeN6EJknqGAqSpI6hIEnqGAqSpI6hIEnqzGRK6haT5NnAvjRbfn6kqr40cEmStKj0fqWQ5Pgka5Jcvk773km+n+TKJEcDVNVnq+ow4Aia1VglSVvQlug+OgHYe7IhyVLgg8A+wG7AQUl2m3jJm9vnJUlbUO+hUFXnAjeu07wncGVVXVVVvwY+AeyXxruAM6vq233XJkm6o6EGmh8A/Hji+Jq27ZXAM4DnJTliui9McniSVUlWXXfddf1XKkmLyLwaaK6q9wPv38hrVgIrAZYvX15boi5JWiyGulK4lmahvSk7tm0zkmRFkpU33XTTnBcmSYvZUKFwEfCQJLskuQtwIHDaTL/YndckqR9bYkrqycAFwK5JrklyaFWtBY4EzgJWA5+sqiv6rkWStGG9jylU1UHraT8DOGM250yyAlixbNmyzSlNkrSOUS5zYfeRJPVjlKHgQLMk9WOUoeCVgiT1Y5ShIEnqh6EgSeqMMhQcU5CkfowyFBxTkKR+jDIUJEn9MBQkSZ1RhoJjCpLUj1GGgmMKktSPUYaCJKkfhoIkqWMoSJI6owwFB5olqR+jDAUHmiWpH6MMBUlSPwwFSVLHUJAkdQwFSVJnlKHg7CNJ6scoQ8HZR5LUj1GGgiSpH4aCJKljKEiSOoaCJKljKEiSOoaCJKljKEiSOqMMBW9ek6R+jDIUvHlNkvoxylCQJPXDUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVJn3oRCkj9I8pEkpw5diyQtVr2GQpLjk6xJcvk67Xsn+X6SK5McDVBVV1XVoX3WI0nasL6vFE4A9p5sSLIU+CCwD7AbcFCS3XquQ5I0A72GQlWdC9y4TvOewJXtlcGvgU8A+/VZhyRpZoYYU3gA8OOJ42uAByS5T5IPAXskedP6vjjJ4UlWJVl13XXX9V2rJC0qdxq6gClVdQNwxAxetxJYCbB8+fLquy5JWkyGuFK4FnjgxPGObZskaWBDhMJFwEOS7JLkLsCBwGmbcgJ3XpOkfvQ9JfVk4AJg1yTXJDm0qtYCRwJnAauBT1bVFZtyXndek6R+9DqmUFUHraf9DOCM2Z43yQpgxbJly2Z7CknSNObNHc2bwisFSerHKENBktSPUYaCA82S1I9RhoLdR5LUj1GGgiSpH6MMBbuPJKkfowwFu48kqR+jDAVJUj8MBUlSZ5Sh4JiCJPVjlKHgmIIk9WOUoSBJ6oehIEnqGAqSpM4oQ8GBZknqxyhDwYFmSerHKENBktQPQ0GS1DEUJEkdQ0GS1DEUJEmdOw1dwGwkWQGsWLZs2dClSLP29//8jaFLmDde/qInDl2CWqO8UnBKqiT1Y5ShIEnqh6EgSeoYCpKkjqEgSeoYCpKkjqEgSeoYCpKkzihDwf0UJKkfqaqha5i1JNcBVw9dxwxsB1w/dBELiO/n3PG9nFtjeT8fVFX3ne6JUYfCWCRZVVXLh65jofD9nDu+l3NrIbyfo+w+kiT1w1CQJHUMhS1j5dAFLDC+n3PH93Jujf79dExBktTxSkGS1DEUJEkdQ0GS1BnldpxjkWR7YKup46r6jwHLGbUk9wXeCOzGHd/Tpw1W1Igl2Rd4BHd8L/9quIrGK8ldgecCOzPxO3Ws76dXCj1I8qwkPwR+BJwD/Dtw5qBFjd9JwGpgF+CtNO/pRUMWNFZJPgQcALwSCLA/8KBBixq3zwH7AWuB/zfxZ5ScfdSDJN8FngacXVV7JHkq8KKqOnTg0kYrycVV9dgkl1bV7m3bRVX1uKFrG5up93Di762BM6vqSUPXNkZJLq+qPxy6jrnilUI/flNVNwBLkiypqq8Co771fR74Tfv3T5Lsm2QP4PeGLGjEbmn//lWS+9O8tzsMWM/YnZ/kkUMXMVccU+jHL9pPX18HTkqyhhFfTs4Tb0+yDfA64FjgXsBrhy1ptD6fZFvgGODbQAEfHrakUXsicEiSHwG30nTJ1dQV7djYfdSDJPeg+TS2BHghsA1wUnv1IA0qyV2r6tapxzSDzf851aZNk2Ta8ZiqGsMKzr/DUOhJ+4PykKo6O8ndgaVVdfPQdY1NkjdU1buTHEvzifYOqupVA5Q1akm+XVWP2VibZibJ24BzgfOravQ9AnYf9SDJYcDhNH3eDwYeAHwIePqQdY3U6vbvVYNWsQAk+X2an8W7tWMyaZ+6F3D3wQobv6uAg4D3J7mZptv43Kr63LBlzY5XCj1IcgmwJ/CtqtqjbbusqhbMYNSWlGQp8K6qev3QtYxZkhcDh9BMepgM2ZuBE6rq00PUtVC0oft84PXAvavqngOXNCteKfTj1qr6ddJ8EEtyJ6bp+tDMVNVtSZ4wdB1jV1UnAicmeW5VfWroehaKJB+muanyZzRXCc+jGcAfJUOhH+ck+Uuay/Q/AV4OnD5wTWN3SZLTgFOYmMnlp9tZ+XKS9wJPbo/PAf6qqtz0fHbuAywFfgHcCFxfVWuHLWn27D7qQZIlwKHAn9L0254FfLh8s2ctyT9N01xV9d+2eDEjl+RTwOXAiW3TnwOPqqrnDFfV+CV5OPBfaKZKL62qHQcuaVYMBWmRSXJJVT16Y22amSTPBJ5Ec+W1LfBN4OtVdfyghc2S3Uc9SPIQ4H/zu4u3/cFgRY1ckocCxwH3q6o/TLI78KyqevvApY3RLUmeWFXfAGjHa27ZyNdo/famGUt4X1X936GL2VxeKfQgyTeA/wX8LbACeAmwpKr+56CFjViSc4CjgH+YmNG1oNac2VKSPJqm62ibtunnwIur6tLhqhq3JPcDptbhurCq1gxZz+Zw7aN+3K2qvkwTuldX1VuAfQeuaezuXlUXrtM22sG8gV1WVY8Cdgd2r6o9DITZS7I/cCHNarPPB76V5HnDVjV7dh/149Z2sPmHSY4ErgW2Hrimsbs+yYNpp/a2/+l+MmxJo/XDdrD5+KpavdFXa2PeDDxu6uqg3fvjbODUQauaJa8U+vFqmjtEXwU8lmZ2x4sHrWj8XgH8A/CwJNcCrwGOGLak0XoU8APgI0m+meTwJPcauqgRW7JOd9ENjPh3q2MKGoUku1TVj9rFBpdU1c1TbUPXNmZJ/hj4OM2smVOBt1XVlcNWNS5JjqHpiju5bToAuLSq3jhcVbNnKPSgnSlzFM1uVpPb87l15CytZxG3i6vqsUPVNFbtsiH70kyA2Bn4GM3Odk8C/rqqHjpcdeOU5LnA1F33X6+qzwxZz+ZwTKEfp9AsgPePwG0D1zJqSR5Gs5fwNkkmb666FxPTfbVJfgh8FTimqs6faD81yZPX8zXagHbZkAWxdIih0I+1VXXc0EUsELsCz6Tp3lgx0X4zcNggFY1Ye5Vwwvo2lXcp8k3Xflh5F7A9zQoGU5vsjHKcxu6jOZRkanvIVwFrgM/Q7MQEQFXdOERdC0GSx1fVBUPXsRAkubCq9hy6joUiyZXAioUyk8tQmEPtdnzF7evUTyrvaJ69JCcCr66qX7TH9wbe49pHmy7J3wJ3Bv4Pd1xccLQrew4pyXlVtWBW8TUUNApJvjN1J/OG2rRxSb46TXM5EWJ2krwP+H3gs9yxZ2CUK/g6ptCDJK+g2ZN58lPtQVX198NWNmpLkty7qn4OXVedP7+zUFVPHbqGBeZewK9oVkWeUsAoQ8ErhR6sZxVKP9VuhiQHA39JM7MrNBuZvKOqPjZoYSOVZF+aWV2TCzZOO/isDUvye+uOF475HprR3nU3zy3N1LZrdDM+7jJgPaNXVR8Fnkuzu9VPgecYCLOT5EM0N1i9kiZg96e5p0azc/rkHeHtvgqj3VTLK4UeJPkbYCeaZRkAXgr8uKpeN1xVC0OS7bnjp9v/GLCcUUpyaVXtPvH31sCZVfWkoWsbo/aq6w00NwTuCnwUeGFVXTJoYbNkn2w/jqIJgpe1x/8CfHi4csYvybOA9wD3p5nu+yBgNU0XiDbN1N4Jv0pyf5q1enYYsJ5Rq6ovJLkz8CXgnsCfVdUPBi5r1gyFOdZ2FV1RVQ+juatZc+NtwF7A2VW1R5KnAi8auKax+nySbYFjaDaYL/zQssmSHEu7am9rG+DfgCOTjPZGQENhjlXVbUm+n2Qnuzbm1G+q6oYkS5IsqaqvJvm7oYsaqXdX1a3Ap5J8nqY7bunANY3RqnWOLx6kijlmKPTj3sAVSS7k9puDqqr2G7CmsftF2/d9LnBSkjVM3HilTfLpJM+uqt9U1a3tlOkv0CzzrhmqqhOHrqEPDjT3oF2OuDukWX3ywKqy/3uW2iWzb6GZMfdCmkv1k6rqhkELG6EkhwH/lWZa7wOB04DXV9WXBi1spNo9rt/C7asiT619NMoVDAyFniTZA3gBzXS/HwGfrqpjh61qYUiyHXBD+cM7a+0NlnvTLJ390nVWS9UmSPKvwGtpuo+6VZHH+oHF7qM51O6jcFD753qatWXiHaSzl2Qv4J3AjTSDzR8DtqO5w/ngqvrikPWNSZK/mDykmTZ9CbBXkr2q6r3DVDZ6N1XVmUMXMVcMhbn1r8DXgWdO7V6V5LXDljR6H6C5k3kb4CvAPlX1zXafhZMBQ2Hm7rnO8afX065N89V297VPc8e1j0a5wKDdR3MoybOBA2l2YPoi8Angw1W1y6CFjdjkkiFJVlfVwyeec+kQDW5igcGpX6ZTYwqjXGDQK4U5VFWfBT7bDoruR7O5/PZJjgM+40DerPx24vEt6zznJ5pZaH+J/c57N9ZfYvPA16ZpG+3PplcKPWun++0PHFBVTx+6nrFJchvN1NMAd6NZjZL2eKuquvNQtY1Vksmpp1vRrCm1tqreMFBJo5ZkcvmarWh2Clw91r0+DAVJ7sY2h5LcFTirqp4ydC2zYfeRtMhMbBsLzX0fj6UZyNfcuDuw49BFzJahIC0+F3P7trFrae6jOXTQikYsyWXcPoawFLgvMNq9Kew+kqTNkGRyL4q1wM+qau1Q9WwuN9mRFokkb5h4vP86z/31lq9oYaiqqyf+XDvmQABDQVpMDpx4/KZ1ntt7Sxai+ctQkBaPrOfxdMdapAwFafGo9Tye7liLlAPN0iLhjYCaCUNBktSx+0iS1DEUJEkdQ0GS1DEUpHkmyS97Pv+zkhzd57+h8XKgWZpnkvyyqrYeug4tTl4paN5LcnCSS5N8N8nHkuyc5Ctt25eT7NS+7oQk709yfpKrkjyvbd8hyblJLklyeZInte1/muSCJN9OckqSrdv2f0/y1rb9snbrT5L8cXuOS5J8J8k9kzwlyecnav1AkkPax+9M8r22zr/ZwPe3S1vHZUnePtGeJMe0NV+W5IC2/SlJzknyufb7fGeSFya5sH3dg9vXrUjyrbbWs5Pcr20/JMkHNvSeafEyFDSvJXkE8GbgaVX1KODVwLHAiVW1O3AS8P6JL9kBeCLNRifvbNteQLO+/aOBRwGXJNmuPe8zquoxwCpgcmP769v244DXt22vB17RnudJ/O5OcJN13wf4M+ARbZ1vX99rgfcBx1XVI4GfTLQ/B5iq+RnAMUl2aJ97FHAE8HDgz4GHtvshfBh4ZfuabwB7tVuWfgJY3yY6071nWqQMBc13TwNOqarrAarqRuDxwMfb5z9G8wttymer6rdV9T3gfm3bRcBLkrwFeGRV3QzsBewGnJfkEuDFwORql1Ob2l8M7Nw+Pg94b5JXAdtuZOGzm4D/BD6S5DncfqPYdJ4AnDzx/Ux5InByVd1WVT8DzgEeN/U9VdVPqupW4N+Aqa1eL5uod0fgrHZp56OAR6zn35/uPdMiZShoobl14nEAqupc4MnAtcAJSQ5un/uXqnp0+2e3qjp0mvPcRrvvSFW9E/jvNHcDn9d2K63ljv+PtmpfuxbYEziV5hP4FzdS96YO7k1+n7+dOP4tt++TcizwgfYK5KVTtW3kXK6BtMgZCprvvgLs33bHTO0adj63r/j5QuDrGzpBu979z6rqH2m6Vx4DfBN4QpJl7WvukeShGznPg6vqsqp6F83Vx8OAq4Hdktw1ybbA09vXbg1sU1VnAK+l6e5Zn/PW+X6mfB04IMnSJPelCbYLN1TjOrahCUJoroSkjXLnNc1rVXVFkncA57Rr93yHps/8n5IcBVwHvGQjp3kKcFSS3wC/BA6uquvaAeGT2z11oRlj+MEGzvOaJE+l+TR+BXBmVd2a5JPA5TQ7mH2nfe09gc8l2Yrm0/dfTHfC1quBjyd5I/C5ifbP0HSVfZfmSuINVfXTqYHvGXgLcEqSn9OE6y4z/DotYk5JlSR17D6SJHXsPpK2kCT/A9h/neZTquodQ9QjTcfuI0lSx+4jSVLHUJAkdQwFSVLHUJAkdQwFSVLn/wPB7yraa+wctwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.barplot(data = consensus_taxonomy_votus.groupby('consensus_domain',as_index=False).agg({'votu':'count'}),\n",
    "           x='consensus_domain',\n",
    "           y='votu',\n",
    "           palette='Set2')\n",
    "plt.xticks(rotation=90)\n",
    "plt.yscale('log')\n",
    "plt.show()\n",
    "\n",
    "# consensus_all_crispr_hits_filter['consensus_domain']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Created curated list of taxonomies to use in analyses: \n",
    "1. want sub proteobacteria\n",
    "2. do not take phylum with \"group name\" in it  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "array(['Alphaproteobacteria', 'Actinobacteria',\n",
    "       'delta/epsilon subdivisions', 'Gammaproteobacteria', 'Archaea',\n",
    "       'cellular organisms', 'Firmicutes', 'Terrabacteria group',\n",
    "       'Betaproteobacteria', 'Bacteria', 'unknown', 'Proteobacteria',\n",
    "       'Planctomycetes', 'Verrucomicrobia',\n",
    "       'Bacteroidetes/Chlorobi group', 'other bacterial phylum',\n",
    "       'Acidobacteria', 'CPR', 'Chlamydiae', 'Gemmatimonadetes',\n",
    "       'Eukaryota'], dtype=object)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "rachel_colors = {'Acidobacteria':                 '#A54657', #(reddish mauve)\n",
    "'Actinobacteria':                '#FFCC00', #(warm yellow)\n",
    "'Bacteroidetes':                 '#F7EE7F', #(cool yellow)\n",
    "'Archaea':      '#4DAA57', #(green)\n",
    "'Eukaryota':                   '#F9ECCC', #(cream)\n",
    "'Firmicutes':                    '#F1A66A', #(peach)\n",
    "'Gemmatimonadetes':              '#F26157', #(pink coral)\n",
    "'Planctomycetes':                '#679289', #(dark sea green)\n",
    "'Proteobacteria':               '#86BBD8', #(light blue)\n",
    "'Verrucomicrobia':               '#33658A', #(dark blue)\n",
    "'unknown': 'lightgrey',\n",
    "'other bacterial phyla': 'dimgrey'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 931,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bacterial_phylum(string):\n",
    "    '''takes row and provides taxonomies to use in analyses. \n",
    "    These are top (most abundant) sub-domain / phyla of bacteria '''\n",
    "    \n",
    "    if 'Candidatus Omnitrophica' in string:\n",
    "        return \"other bacterial phyla\" \n",
    "    \n",
    "    elif 'Terrabacteria group' in string:\n",
    "        if len(string.split(\",\")) < 3:\n",
    "            return 'Terrabacteria group'\n",
    "        elif string.split(\",\")[2] in ['Chloroflexi','Deinococcus-Thermus', \n",
    "                                          'Cyanobacteria/Melainabacteria group',\n",
    "                                          'Tenericutes', 'Armatimonadetes']:\n",
    "            return 'Terrabacteria group'\n",
    "        else: return string.split(\",\")[2]\n",
    "    \n",
    "    elif \"Eukaryota\" in string:\n",
    "        return \"Eukaryota\"\n",
    "    elif \"arch\" in string.lower():\n",
    "        return \"Archaea\" \n",
    "    \n",
    "    \n",
    "    elif \"Proteobacteria\" in string:\n",
    "        if len(string.split(\",\")) < 3:\n",
    "            return \"Proteobacteria\"\n",
    "        else: \n",
    "            # classify oligoflexia and acidithiobacillia as Proteobacteria ... \n",
    "            if string.split(',')[2] in ['Acidithiobacillia','Oligoflexia']:\n",
    "                return \"Proteobacteria\"\n",
    "            else: return string.split(',')[2]\n",
    "    \n",
    "    elif \"Acido\" in string:\n",
    "        return \"Acidobacteria\"\n",
    "    \n",
    "    elif \"Bacteria incertae sedis\" in string:\n",
    "        if len(string.split(\",\")) < 3:\n",
    "            return 'other bacterial phylum'\n",
    "        elif \"Microgenomates\" in string:\n",
    "            return \"CPR\"\n",
    "        elif \"Saccharibacteria\" in string:\n",
    "            return \"CPR\"\n",
    "        else: return 'other bacterial phyla'\n",
    "        \n",
    "    elif len(string.split(\",\")) < 2:\n",
    "        return string\n",
    "    \n",
    "    elif \"group\" in string.split(\",\")[1]:\n",
    "        if len(string.split(\",\")) < 2:\n",
    "            return string\n",
    "        if len(string.split(\",\")) < 3:\n",
    "            return string.split(\",\")[1]\n",
    "        else: return string.split(',')[2]\n",
    "    \n",
    "    elif string.split(',')[1] in ['Calditrichaeota','Chrysiogenetes','Atribacterota','Dictyoglomi',\n",
    "                                 'Elusimicrobia','Synergistetes','Deferribacteres',\n",
    "                                 'Thermodesulfobacteria','Thermotogae','Nitrospirae','Fusobacteria',\n",
    "                                 'Aquificae','Duplodnaviria','Spirochaetes']:\n",
    "          return \"other bacterial phyla\"\n",
    "    \n",
    "    \n",
    "    elif 'cellular organisms' in string:\n",
    "        return \"unknown\"\n",
    "    \n",
    "    else: return string.split(',')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 932,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count              24903\n",
       "unique                21\n",
       "top       Actinobacteria\n",
       "freq                8414\n",
       "Name: sub_domain_consensus, dtype: object"
      ]
     },
     "execution_count": 932,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_taxonomy_votus['sub_domain_consensus'] = consensus_taxonomy_votus[\"taxa\"].apply(\n",
    "    lambda x: get_bacterial_phylum(x)).str.replace(r'^\\s*$','unknown',regex=True).str.replace(\n",
    "    'cellular organisms',\"unknown\")\n",
    "consensus_taxonomy_votus['sub_domain_consensus'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 933,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Alphaproteobacteria', 'Actinobacteria',\n",
       "       'delta/epsilon subdivisions', 'Gammaproteobacteria', 'Archaea',\n",
       "       'unknown', 'Firmicutes', 'Terrabacteria group',\n",
       "       'Betaproteobacteria', 'Bacteria', 'unclassified bacterial viruses',\n",
       "       'Proteobacteria', 'Planctomycetes', 'Verrucomicrobia',\n",
       "       'Bacteroidetes/Chlorobi group', 'other bacterial phyla',\n",
       "       'Acidobacteria', 'CPR', 'Chlamydiae', 'Gemmatimonadetes',\n",
       "       'Eukaryota'], dtype=object)"
      ]
     },
     "execution_count": 933,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_taxonomy_votus['sub_domain_consensus'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 936,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_taxonomy_votus = consensus_taxonomy_votus.replace(\n",
    "    {'sub_domain_consensus':{'unclassified bacterial viruses':'Bacteria'}})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 944,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_taxonomy_votus.to_csv(\n",
    "'/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/host_info/consensus_host_taxa_3.csv',\n",
    "                                       index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_taxonomy_votus = pd.read_csv(\n",
    "'/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/CRISPRs/host_info/consensus_host_taxa_3.csv' \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "lazy_coloring = {}\n",
    "for taxa in consensus_taxonomy_votus['sub_domain_consensus'].unique():\n",
    "        if \"arch\" in taxa.lower():\n",
    "             lazy_coloring[taxa] = rachel_colors[taxa]\n",
    "        elif \"Eukaryota\" in taxa:\n",
    "            lazy_coloring[taxa] =  rachel_colors[taxa]\n",
    "        elif \"proteobacteria\" in taxa.lower():\n",
    "            lazy_coloring[taxa] = rachel_colors['Proteobacteria']\n",
    "        elif \"delta\" in taxa.lower():\n",
    "            lazy_coloring[taxa] = rachel_colors['Proteobacteria']\n",
    "            \n",
    "        elif \"Bacteroidetes\" in taxa:\n",
    "            lazy_coloring[taxa] = rachel_colors['Bacteroidetes']\n",
    "            \n",
    "        elif \"Bacteria\" in taxa:\n",
    "            lazy_coloring[taxa]='dimgrey'\n",
    "        elif \"CPR\" in taxa:\n",
    "            lazy_coloring[taxa]='#9ebd8f'\n",
    "        elif \"Terrabacteria group\" in taxa:\n",
    "            lazy_coloring[taxa] ='#f17e6a'\n",
    "        elif \"Chlamydiae\" in taxa:\n",
    "            lazy_coloring[taxa] = '#7494ab'\n",
    "        \n",
    "        else: lazy_coloring[taxa] = rachel_colors[taxa]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_domain_order = ['Actinobacteria',\n",
    "    'Alphaproteobacteria',  'Betaproteobacteria', 'Gammaproteobacteria', 'delta/epsilon subdivisions', 'Proteobacteria',\n",
    "    'Firmicutes', 'Bacteroidetes/Chlorobi group','Terrabacteria group', 'Planctomycetes', 'Verrucomicrobia', \n",
    "    'Gemmatimonadetes','CPR', \"Chlamydiae\",\n",
    "    'Archaea',  'Eukaryota', 'Bacteria','other bacterial phyla','unknown',  \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZ0AAAGPCAYAAACHwQ0yAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy89olMNAAAACXBIWXMAAAsTAAALEwEAmpwYAABVuElEQVR4nO2dd7hjVdX/P98ZROoAr4AFpSOKFMEBBfUVsMErRQXEEeurIBbELpafIPraURFFRJpSRFFRQBDpIH3ooKBIEVSKKDCiIDDf3x9rZ25u5pZzktzkJrM+z5MnOSfZO+sk55y199qryDZJkiRJ0gtm9FuAJEmSZNEhlU6SJEnSM1LpJEmSJD0jlU6SJEnSM1LpJEmSJD0jlU6SJEnSMxbrtwCdsOKKK3r11VfvtxhJkiQDxRVXXPE32yt12MfKiy222GHA+oyewMwHrn/sscfe8bznPe+e1nYDrXRWX3115s6d228xkiRJBgpJt3fax2KLLXbYU57ylGevtNJK/5gxY8aCgM/58+fr3nvvXe+uu+46DNihtd20Mq9JWlrSXEnb9VuWJEmSZELWX2mllR5sVjgAM2bM8EorrfQAMQNaiClVOpKOkHSPpOtb9m8j6SZJN0vap+mtjwE/nkqZkiRJkq4wo1XhNL1hxtEvUz3TOQrYpnmHpJnAt4FtgfWAOZLWk/Ry4LfAQjbAJEmSZDiY0jUd2+dLWr1l92bAzbZvAZB0PLAjsAywNKGI/i3pVNvzp1K+JEmSpLf0w5FgFeCOpu07gefbfi+ApLcCfxtP4UjaA9gDYNVVV51aSZMkSZLxmD9//nyNZWKbP3++CC+2hZhWjgQAto+yfcoE7x9qe7bt2Sut1JHHX5IkSdI+1997773LFQWzgOK9thxw/ViN+jHT+TPwjKbtp5d9lZG0PbD92muv3U25kiRJkoo89thj77jrrrsOu+uuu8aN0xmrnaa6nk5Z0znF9vplezHg98BLCWVzOfAG2zfU7Xv27NnOOJ0kSZJ6SLrC9ux+fPeUznQk/RDYElhR0p3AvrYPl/Re4HRgJnBEOwqnmdt+vUbtNqu/4tZOvjJJkiRpg6n2Xpszzv5TgVPb7TfNa0mSJIPJtHMkqILtk23vsdxyy/VblCRJkqQGA6l0kiRJksFkIJWOpO0lHfrAAw/0W5QkSZKkBgOpdNK8liRJMpgMpNJJkiRJBpNUOkmSJEnPGEilk2s6SZIkg8lAVg61fTJw8uzZs3fvRn8Hnn5VW+32fuXG3fj6JEmSRYaBnOkkSZIkg0kqnSRJkqRnpNJJkiRJesZAKp10JEiSJBlMBlLpZHBokiTJYDKQSidJkiQZTFLpJEmSJD0jlU6SJEnSMwZS6aQjQZIkyWAykEonHQmSJEkGk4FUOkmSJMlgkkonSZIk6RmpdJIkSZKekUonSZIk6RmpdJIkSZKekUonSZIk6RkDqXQyTidJkmQwGUilk3E6SZIkg8lAKp0kSZJkMEmlkyRJkvSMVDpJkiRJz1is3wIMCweeflVb7fZ+5cYd9dHcPkmSZLqTM50kSZKkZ6TSSZIkSXpGKp0kSZKkZ6TSSZIkSXrGQCqdzEiQJEkymAyk0smMBEmSJIPJQCqdJEmSZDDJOJ0h4/Yffap2m9V2/dwUSJIkSbIwOdNJkiRJekYqnSRJkqRnpNJJkiRJekYqnSRJkqRnpNJJkiRJekYqnSRJkqRnpNJJkiRJekYqnSRJkqRnpNJJkiRJesa0UTqSni3pEEk/kfSufsuTJEmSdJ8pVTqSjpB0j6TrW/ZvI+kmSTdL2gfA9u9s7wm8DnjhVMqVJEmS9IepnukcBWzTvEPSTODbwLbAesAcSeuV93YAfgmcOsVyJUmSJH1gSpWO7fOBv7fs3gy42fYttv8DHA/sWD5/ku1tgd2mUq4kSZKkP/Qjy/QqwB1N23cCz5e0JfBa4IlMMNORtAewB8Cqq646ZUImSZIk3WfalDawfS5wboXPHQocCjB79mxPrVRJkiRJN+mH99qfgWc0bT+97KtMlqtOkiQZTPqhdC4H1pG0hqTFgdcDJ9XpIMtVJ0mSDCZTal6T9ENgS2BFSXcC+9o+XNJ7gdOBmcARtm+YSjmS6tz9p6+31e7Jq36gy5IkSTKMTKnSsT1nnP2n0oFbtKTtge3XXnvtdrtIkiRJ+sC0yUhQhzSvJUmSDCYDqXSSJEmSwSSVTpIkSdIzBlLppMt0kiTJYDKQSifXdJIkSQaTgVQ6SZIkyWAybdLgJMPD7V/5VFvtVvvI5xa83utrX6jd/qAPfryt702SpHcM5Ewn13SSJEkGk4FUOrmmkyRJMpgMpNJJkiRJBpNc00mGlgNPv6p2m71fufEUSJIkSYOBnOnkmk6SJMlgMpBKJ9d0kiRJBpOBVDpJkiTJYJJKJ0mSJOkZqXSSJEmSnpFKJ0mSJOkZA6l00nstSZJkMBlIpZPea0mSJIPJQCqdJEmSZDBJpZMkSZL0jEyDkyTjcP3117fVbv311++yJEkyPORMJ0mSJOkZqXSSJEmSnjGQSiddppMkSQaTgVzTsX0ycPLs2bN377csSTIR7373u9tqd/DBB3dZkiSZHgzkTCdJkiQZTAZyppMkixJvOPJttdsc97Yjp0CSJOmcnOkkSZIkPSOVTpIkSdIzUukkSZIkPWPSNR1J/z3Wftvnd1+cJEmSZJip4kjwkabXSwCbAVcAW0+JREmSJMnQMqnSsb1987akZwDfmCqBkiRJkuGlnTWdO4Fnd1uQOmRGgiRJksGkyprOQYDL5gzgucCVUyjTpGRGgiRJksGkyprO3KbXjwE/tH3hFMmTJEmSDDFVlM7ytg9s3iFp79Z9SZIkSTIZVZTOW4BWBfPWMfYlSTJNOfD0q2q32fuVG0+BJMmizrhKR9Ic4A3AGpJOanprFvD3qRYsSZIkGT4mmulcBPwVWBE4oGn/PODaqRQqSZLpRZZoSLrFuErH9u3A7cDmkp4MbFre+p3tx3ohXJIkSTJcTBqnI2kX4DJgF+B1wKWSdp5qwZIkSZLho4ojwaeATW3fAyBpJeBM4CdTKViSJEkyfFTJSDCjoXAK91VslyRJkiSjqDLTOU3S6cAPy/auwKlTJ1KSJEkyrFRROvcAxxDpbwAOtX3ilEmUJEmSDC1VzGRLA/sQJQ1uJVypkyRJkqQ2kyod25+x/RzgPcBTgfMknTkVwkh6taTvSfqRpFdMxXckSZIk/aOOQ8A9wF2EI8HKVRtJOkLSPZKub9m/jaSbJN0saR8A2z+3vTuwJ7F2lCRJkgwRVeJ03i3pXOAs4EnA7rY3rPEdRwHbtPQ5E/g2sC2wHjBH0npNH/lUeT9JkiQZIqo4EjwDeL/tq9v5AtvnS1q9ZfdmwM22bwGQdDywo6TfAV8ETrM9Zs0eSXsAewCsuuqq7YiUJEmS9Ikqazofb1fhTMAqwB1N23eWfXsBLwN2lrTnOPIcanu27dkrrbRSl8VKkiRJppIqM52eYfubwDf7LUeSJEkyNfQrs8CfCbNdg6eXfZWQtL2kQx944IGuC5YkSZJMHf1SOpcD60haQ9LiwOuBkyZpswDbJ9veY7nllpsyAZMkSZLuM+VKR9IPgYuBdSXdKentpTTCe4HTgd8BP7Z9w1TLkiRJkvSXKV/TsT1nnP2n0mYON0nbA9uvvfbanYiWJEmS9JiBzBad5rUkSZLBZCCVTpIkSTKYDKTSSe+1JEmSwWRaxelUxfbJwMmzZ8/evd+yJElSjZe8Y7/abc47rH6bZHozkDOdJEmSZDBJpZMkSZL0jIE0r6XLdJIsmhy601trt9njp0d1XY6kfQZyppMu00mSJIPJQCqdJEmSZDBJpZMkSZL0jIFUOhmnkyRJMpgMpCNBxukkSdIu3//VwbXbvGWbd0+BJIsmAznTSZIkSQaTVDpJkiRJz0ilkyRJkvSMVDpJkiRJzxhIpZPea0mSJINJeq8lSZLU4LYPvaetdqsf8O0uSzKYDORMJ0mSJBlMUukkSZIkPSOVTpIkSdIzUukkSZIkPSOVTpIkSdIzBlLppMt0kiTJYDKQSieLuCVJkgwmA6l0kiRJksEklU6SJEnSM1LpJEmSJD0jlU6SJEnSM1LpJEmSJD0jlU6SJEnSM1LpJEmSJD0jlU6SJEnSMwZS6WRGgiRJksFkIJVOZiRIkiQZTAaycmiSJMkg86lvHVu7zefeu9sUSNJ7BnKmkyRJkgwmqXSSJEmSnpFKJ0mSJOkZqXSSJEmSnpFKJ0mSJOkZqXSSJEmSnpFKJ0mSJOkZqXSSJEmSnpFKJ0mSJOkZqXSSJEmSnpFKJ0mSJOkZ00bpSFpT0uGSftJvWZIkSZKpYUqVjqQjJN0j6fqW/dtIuknSzZL2AbB9i+23T6U8SZIkSX+Z6pnOUcA2zTskzQS+DWwLrAfMkbTeFMuRJEmSTAOmVOnYPh/4e8vuzYCby8zmP8DxwI5TKUeSJEkyPejHms4qwB1N23cCq0h6kqRDgI0lfXy8xpL2kDRX0tx77713qmVNkiRJusi0KeJm+z5gzwqfOxQ4FGD27NmearmSJEmS7tGPmc6fgWc0bT+97KuMpO0lHfrAAw90VbAkSZJkaumH0rkcWEfSGpIWB14PnFSnA9sn295jueWWmxIBkyRJkqlhql2mfwhcDKwr6U5Jb7f9GPBe4HTgd8CPbd8wlXIkSZIk04MpXdOxPWec/acCp7bbr6Ttge3XXnvtdrtIkiRJ+sC0yUhQhzSvJUmSDCYDqXSSJEmSwWQglU56ryVJkgwmA6l00ryWJEkymAyk0kmSJEkGk1Q6SZIkSc8YSKWTazpJkiSDyUAqnVzTSZIkGUwGUukkSZIkg0kqnSRJkqRnpNJJkiRJesZAKp10JEiSJBlMBlLppCNBkiTJYDKQSidJkiQZTFLpJEmSJD0jlU6SJEnSMwZS6aQjQZIkyWAykEonHQmSJEkGk4FUOkmSJMlgkkonSZIk6RmpdJIkSZKekUonSZIk6RmpdJIkSZKekUonSZIk6RmL9VuAdpC0PbD92muv3W9RkiRJ+sI/772udptlVtpgCiSpx0DOdDJOJ0mSZDAZSKWTJEmSDCapdJIkSZKekUonSZIk6RmpdJIkSZKekUonSZIk6RmpdJIkSZKekUonSZIk6Rmy3W8Z2kbSvcDtE3xkReBvHX7NsPQxHWSYLn1MBxmmSx/TQYbp0sd0kKEbfVRpv5rtlTr4jrYZaKUzGZLm2p6dfUwPGaZLH9NBhunSx3SQYbr0MR1k6EYf3ZBhKknzWpIkSdIzUukkSZIkPWPYlc6h2ce0kmG69DEdZJgufUwHGaZLH9NBhm700Q0ZpoyhXtNJkiRJphfDPtNJkiRJphGpdJIkSZKekUonSZIk6RmpdJJxkbSmpJMl/U3SPZJ+IWnNNvuaJWnZbstY8bt3aXy3pE9J+pmkTfohy7AgaQVJm0n678aj3zIlg8HQORJIehXwHGCJxj7b+/dJlhWAdVpkOb9G+46ORdKXgc8B/wZ+BWwIfMD2MRXbXwJ8G/hh2fV6YC/bz68hw6bAEcCygID7gf+1fUXVPko/LwSutv2QpDcCmwAH2p4oI0Wj7bW2N5T0IuL3+Arw6TrHUfp5IrATsDpNpd77cX5JWpnR58Wfevjd7wD2Bp4OXA28ALjY9tY1++n0/F4CePsYffxvHTnG6PeFti+s+NmlgX/bni/pmcCzgNNsP1rj+5YCPgSsant3SesA69o+pabc0+beNxFDNdORdAiwK7AXcYPbBVitZh8rSfqqpFMlnd14tCHLO4DzgdOBz5Tn/Wq07/hYgFfYfhDYDrgNWBv4SI32S9k+2vZj5XEMTSd0RQ4H3m17ddurAe8BjqzZB8B3gH9J2oi4QP8I/KBi28fL86uAQ23/Eli8DRl+AewIPAY81PSoTKfnl6QdJP0BuBU4j/hfT6spwzqSfiLpt5JuaTxqdLE3sClwu+2tgI2JwUQdGbpxfh8NPAV4JfFbPB2YV/H7Z0qaI+nDktYv+7aTdBHwrRoynA8sIWkV4NfAm4CjarSHuB4eATYv238mBkeV6dLv2RtsD80DuLbleRnggpp9/JoYPf0OeAkxSv9SG7JcR9ygry7bzwJ+1uNjub48HwZsU15fU6P9l4B9iJH9asBHgS8A/wX8V8U+rhpj35Vt/J5XludPA2+v0w9wCvBd4BZgeeCJdX6H1t+zw3O0o/MLuAZ4UuN3BbYCDq8pw2+AlwLXlv91P2D/Gu0vL89XA08sr2+oKUM3zu+rWvp4AnBJxbZHAWeV8/ls4BjgRuDVbZ6XewEfbfwuNfuY23w8jf+5179nrx4LTARDwr/L878kPQ24D3hqzT6eZPtwSXvbPg84T9LlbcjysO2HJSHpibZvlLRujfbdOJZTJN1Y+nqXpJWAh2u0f115fmfL/tcDBqqs75wn6buEic7EaOzcxpqK7SsryjJP0seJkeSLJc0gbjJVeB2wDfBV2/dLeir1ZnwNLpK0ge3r2mjboNPz61Hb90maIWmG7XMkfaOmDEvaPkuSHObJ/SRdQSj0KtwpaXng58AZkv7BxIl3x6Ib53fDhHV/ma3cBaxcse1sYEOHWWyJ0nYt2/fVlEGSNgd2IwYTADNr9vEfSUsS1weS1iJmPnXoxu/ZE4ZN6ZxSLoavAFcSf+JhNftonMh/LTbSvxAj+7p0emF2fCy29ynrOg/YflzSvwjzUNX2a9T5vnHYqDzv27J/Y+KYqq4D7Aq8gVgPukvSqsRvMym2/yXpF8CTSzuIUW1dXgS8VdKtxE1B0b03rNFHp+fX/ZKWAS4AjpV0DzVNfMAjRWn/QdJ7CXPOMlUb235NebmfpHOA5Yg1wzp041o9tKyb/j/gJOIYqirO/9ieD1AGh7e0oXAA3g98HDjR9g3F0eacmn3sS/x+z5B0LPBC4K01++jG79kThs6RoEFZ9F3C9gM1221HXNDPAA4CZgGfsX1SB7K8hHJh2v5PG+3bPZalgA8SC5R71F2glPTmsfbbrrqW0lUkrQasY/vMcmwzbU9qw5e0F3Fh3w3ML7vrKovG9y+EKzgzNPXR0fnVWLgm1mN3I86rY+vcMItzx+8IU+NnSx9ftn3JJO0mVI62/15VhpZ+2zq/O6EMwG5ubAJrle12BhJIWsr2vzqQ50mEQ4YIE2Gt0gbFmvJI4zVh2n+4sW86MRRKR9LWts+W9Nqx3rf9sx7KMsv2g+NdoJNdmN08Fkk/Aq4A3mx7/XKjvsj2cyu2P6hpcwliHeBK2zvXkGHMkadretVI2h3Yg1hLWqso0ENsv7RC25uB57c5km3u57PEwvFFtuvOLrpGu8q3C997KzGCVtPuxrZtT2pu7cb5LemNto+R9MFx+vhahT4mXGSvOpAoprXDgWVsr1ocXd5p+91V2jf106mn65W2N5ls33RgWMxrLyEWA7cf4z0DVU7kj9r+crnRLqSJbb+voizHEd5iV9B0QTY9T3ZhdnwsTaxle1dJc2CBmUmTNVrwZfZezdtl+n58je+H0aafJYjf5nc1+4DwetsMuLTI9geF23AV7gC6MYq+BZgDfFPSPGLGcr7tX0zWsFvnV7PyJUbnqwCHEAOCydp+w/b7JZ08jgw7TNS+S+bWbpzfS5fntuO+bN8u6dWER+d1tk9vs6tvEN5zJ5V+r1HNmCWN44JOBdOzpKcQ58CSkjZmZEAwC1iqjhy9YiiUju19i436NNs/brObxo1wboeybFdu7C9xG7ETXTqWBt1YoGzmIaDWjcf2Ac3bkr5KuI/X5RHb/2noTEmLMcaNcxxuIZwXfknT8VcZETdj+0jgyHKhvw74MKEAqtz8unJ+0ZnyPbo8f7UTAST9lBjd/6qxLlIV2/uW57e1+/22vytpJvCg7a+304ekg4mYlouAz0razPZn25Tnjpax3OPjfXYcGi7ol9jeStKzgM9XbPtKYv3n6UDz+TwP+ERNOXrCUCgdgOKF8lGgrRu17ZPLibyB7Q93KIvLDW6DNtt3dCxN7MfCC5SVL/aWEfFM4NldkGkp4gKpy3mSPkGM6F4OvBs4uWLbP5XH4rQXnwOApMOA9Yi1oQuAnYlF20np4vnVtvK1fUWRYQ/bu3Ugw3eI8+ggSScAR9q+qU4HZQ1jX8I5w4Qb9/5VTaDFMWYO0JbSAf4b2Kj0sxTxf7ajdO6QtAVgSU8gFEjdmXzbnq62vw98X9JOtn9a83v7wtAoncKZkj4M/Igms07VBc5yAr6wS7JcKWlT2+24W0OHx1I++2uFK2xjgXLvmguUzSPix4hgwDtrtEfSdYxWXCsB7URJ70O4pF5HuHCfavt7VRra/kyRZZmy/c82vh8iPmYmEQj5d+Bvth+r2rhL51cnyrchw2qSFm/HqaX0cSZxfi5HmBvPlHQH8D3gGFeLxj+eWB/bqWzvRpzrL6shyoWSvsXC10iVgcB/bD9ePl/L7NzCnsCBhInrz0QcVq31HLrjgn6WpK8RyhQiWHb/XjpnVGUoHAkalIXOViotcDb18R3iBDqB0SdyLWcERXzM2sTJ8xA1vWK6dCxntS60j7Vvkj6eTEz9AS6zfU/VtqV984LtY8DddW7UTf3sbfvAyfaN03Z9wrTUcO74G+FccUNdOUp/zybMGh8gFvErz9w6Pb+K6fXtwCuIc+p04DDXuJAl/YCYtZ7UIkNlc2OZqbyRiJv6C3AsMWvZwPaWFdpfb3v9ln3X2a5sHVC4a7diV0jHo/G91xqdVL1OF0qZM9a+qqhNT9di8rwe+H7Z9SZiJjemw0Y/GSql0w0kjZWixa6Zz2k875iqXjGdoAh2W4qIF9iS0YuLv7L9rIr9vI7w+z+39PFi4CO2f1JTno1KW4iF92vrtC99jOWdc5XtjSu0vQj4pO1zyvaWwOdtb1FThu2I4/hvwt34EiLq+4gafXTl/OoESa0xUw0hPlOx/YnAuoQiP8r2X5vem2t7doU+vgZcxoi5dmdgs05N21VReD8+mXAyaeYZwF22b1641Zj9tO01pg49XVv6utotXqlj7ZsODJXSUYdxKVMgz4sI19YjFdkAlrE91gxmrLZtH4ukvYmgtacRo9AGDwLfs10pt5Ska4CXN2Y35RjOtL3RxC0XkmV3RrySXkPkPzto/Faj2s8hgkJfRNjdGywLzK8ya5N0TavMY+2r0M+3igwX2P7LZJ+fCsp58AVibanZvbZ29m+1GVsiaauGAm+j7TxGvDmXZiRuagbwT9uzavT1ZGLB/Wm2t5W0HrC57cMrtD0F+LhbsktI2oAYkIzlXdf8uc2BLYjrrHldaRbwmirnlqRTHI5HY7qi17RqXEwMCH9Ttl9IZODYfOKWfcDTIBdPtx6EbfejjOQcW4r6eZCeSeRkavSxIfCpNmTZl7C1/75sPw24sMfHsleHv+d1LdszWvdV6ONaYOmm7aUp+aEqtl+NmK1dTLjbNh6bAItV7ONEImp99fL4FBFB3s5v8mTC7Xs7YOU22nd0ftFh3rTSx+bAb4E/le2NgINr9rEFMRh4c+PRybnW5n9xGuFFeE3ZXqzq+UnJHzfOe5P2Uc7BfYG/lufG44PEQLPXv8Vzibx8t5XHVUSan57KUeUxbDOdubZnN5td6o5oJZ1H5OX6blMfC9mfK/RzNZHq5cqmfq51dVtxN46lo4wCkr5C3BQbpQ12JRTGx2rIcB2wqe2Hy/YSxAVf27NPo4MilySUTpWMBCsQmb5fVHZdAOxn+x81v38XwrniXNo0N3Z6fkm6wvbzmtc/GvtqyHApYc46qU0ZjibWQK5mxD3Yrh7L1uhnB0YWvs91/VT+l9vetOUaudoVTEqS/mB7nXHeu9n22hVlWM0R81N71qhJajq5el5CJM10OInMKm0frCNLLxk277VuxKUsZfuyFmeW2gvfhHeMJTVkWXqyBq3tu3Asmza9XpBRgAolAYo3zzdLH42b9aG2T6wpw5HApWUdAODVRIxHLbRwUOTTqRgUWZRLrRviOHyKUKCjzI1AnTWuTs+vjvKmNXBnsSWzgfXcwYhV0heJc+vYsmvvsgD/8RrdPFQcGhrXyAuoHgQ8V9LubvGAVARq1qn19DRJpxH/Qd2MBI0YtiWI3/QaYjCzIRHPVcc09ofiTHCE7XaCr3tHv6da3XwQHj3nAfcSJ/NtwFY1+ziNuKk1UpbvTARq1pXlw4yk09+dMA+9r5fHMkafyxOOBFU/X8uUNkb7GYQZZhPipv8+YOM2+7qaiLG5qqp8wDfK88mEp9aoRxsydMPc2NH5RdyolyGU7pHEWtkLasrwk/K/XElk6v4wcHyN9icAT+3w3LgWmNG0PZMaZtfSZhPgQkLRXAj8nvDYqtL2yURg6LnEzf+Acr1dDDylhgyXEs4HzedlrRIY5T/coGl7feAnNftYttxnLiIcXPYAZnXyH03VY6jMa7DAlbOTxHlrAocSF+U/iGJZu7kNrzNFHMUC11bbZ9Rs39GxjNHfE4gLolLgmaTvA99y+7FGlT3MKvRzqe3nN/pTBEVe6QnMlZKe5wiIfMlY7ztKC9SRoRvmxq6dX+0iaUUituRlxLn1a2JANFlewEaw8LLEGsJljM7wMGEanZa+rgW2bHxn8eA6d6L/c4w+nkjM0NYtx3ETocgqWwQkbUXc5CFqAtUq2Nh6XpZ9dc3gN9h+zmT7avT3EiId1/LEAOOzruiN1wuGyrymkRiUX46xryq2/bJiDpthe56k2jmnJH2p3IzOGGNflfYdH4tGZxSYQXg81cko8HxgN0ltxRoVzpK0E1HArpMRznmqGRTpUhK7WbmU9Z1nuA23bdsfKcfSCPBsx9zY0fmlKIn8EcKJoLlkdp1S0eu6JSNB8XaaLLako/Q5LXweuEoRayNibWefmn1c7HBNXhBvJelKYgZUCYcXXlueeIVuZCS4VpHtolFGfjdiJlgZRaaJVxGZIlYnZm7HEuuOpxIOLNOCoZjpqEtxKaWvsfzuay3UTtDPpI4EXT6W5hF+7YwC6k4q/3mEx9pjRAG5huKq7Bpb+mk7KFLSucAOxE36CuAewpNwzCzFU0mn55fCjf0Q4jgWrMM0FGwHMtTKSKzIP7cZMai53PZdNdrOIMyKFzA68LhSHxpJcnkM4UHXfI0cUuca6ZRxZo17u16piSWAdzHiVHE+8B0X55uKfdxC3DMOt31Ry3vfdE0nj6lkWGY672QkLuUKRk7CB6lY71yRZO85wHIanXZ9Fk3xEBX6eRcxCl+rmBAaLEvYWyej42Np4k/AXz3iObakpNVt31ax/VieYbVS6NtuOxNwC0sSi6TfgwUjuyWBKh5DyzmC8N4B/MCRVLWdANXXEiW8Vyb+l8oKtFvnF/CY7e/U+HyzDI3YkpU0uizALGpUuyy/46eJbNEicrDt74pBsi65BR0JbdupUzVekssH6XGSy2Ly7iSPHcDLgQPdfvLSmUSQ7pjppaaTwoEhmek0kLSXKwYdjtF2R8KzagdGXwjziEXWKgoDRT6qFYgAvmZzwbzJbOYt/bR9LE19zAW2cEmnIWlxYoS/6cQtF7S/jVgk/Qdxc1meKOt7N7B7ldG1xnYLfYCYdVX22pJ0CfAyl7xpijxqv3aFrAIKt+1XEClCPmn78iqzzjH6uRnY3m14B3V6fmkkav19xEztREavp0x6bpWZ75ZEvrBDWmQ42fYfJuuj9HMTcV7dV7afRNQYqlyOvXiv/Y0Ocguqj0kuNU6JigZ1bvSSjiE81RreZ7Wr2kq6zPZmddv1g2GZ6TSYL2l52/fDAvv9HNsHT9bQURPlF5I2t31xuwI4Euw9IOlA4O8ucSSSZkl6vu1LK3bV9rE0sZib8jc5shPXybJ8BuFFc3qR4RVEgsYjgYOJNZ/JOJiwsTcivzcgckQtJ+ldtn9dUZYl3JSo0/Y/FVkbqrA/YY77TVE4awKVbrAt3N2OwoGunF/N9Zkg1nUWdM/kdZoaa1vnSTqqQ8eF+xg9451X9tVh1/L8nmYRqXAcTTyvrHPeDwuukQ/Z/lRNWdqhUaLihcRa6Y/K9i5E4G1lbL9REV8zBzhKEWZxJPBDVy/O10ny097iaeBC160HY0Ts0+TKWLGP7wPLN22vQIw+6spyFWUmWbZnUNxke3gsZwA7NG3vCJxVo/1C7sAUt9ax5Bunj58Bz2naXo/wqFmzah+l3YXAJk3bzyMWknt5fh1IXNRzgNc2Hv04vzo8jjPGkOH0Gu1/UM7v/Ygo/CuBo4ho/A/28DiuGmNf5WusSzJcQlNmDMIF/ZI2+3oSYVq/jXCt/wMVs4oQ6zmtj7N7+VtUfQzbTGemJLn8C8XWWbd+yoYuIyeIwEJFRb66LJCj9DNf4eZblW4cy57AsWUEJCK54ZhZCsbhr5I+xki10F2Bu4ssVYt3PdNN2Zxt/1bSs2zfonrZ5N8PnCDpL8SxPIWR0fKYqHvVYBvMItaQXtHcDfWquXZ0fkl6D3CsO5sBrziGDFULwQH8sTwa/KI8V16/69JxzFTUn3mk9LEk8MQa7bvBCsR50TALLlP2VUaRmeFtRFb6HxCJT+8pM/nfApOa2W1vVec7+8mwKZ1fAT+S9N2y/c6yrw4zJK3gkiKl2NLb+Z1ukfQ+ouAVhHPBLTXad3wstv8IvEDt15F5AzGS/Tlxc72w7JtJ5Lyqwg2KdP7Niuu3ihiLKnVXAHCYxZ5FxGQA3OTJ67Z0q1pngw+5Zc1B9d3pOz2/drf97cZGURi7E2bMqsyXtKpLZdvipVh5cdcVs1FPQjeO41jCJf/Isv02RlL794ovsrDr9341+9gJ+Lrt85t3Our8vL1qJ5JeRTirNCeCbad21ZQybI4EM4ibcyOW5QzCrbZyig9FvrJPEFHXIlw7/8/20RM2XLiflYk0MlsTF/RZwPtdsR5NJ8ci6Y22j2nxUFqAa5Zp7oQy+nw3I6l0LiRuLA8TKWEqKUJFMbrDCTv3P2rKsIm7YNuWdCGwrUteK0VdnRNcIy9fp+dXcYrYsGUGfK1rBBJK2oYIUD2vyPBioppoO2XE26Ibx1HabUvTNdLLY2iS4SmMrG9e6hru4y39zGJ07FUdp4pDiFCLrYDDiPPqMtuVlVavGCql0y0kPYf48yDsorUWBvuNpHc66sjvO9b7XRqp9hRJaxMj2V2JmcuRhPdalTidcwhz3E+AH9m+vk0ZXkVk/n4VMeP6AZFN4Oqa/bR9fkn6KrAqkWIJYmByh+0P1ZRhRSLbBXQh20VdFNkdVqPD45gOFNPgOoyeYZw/fouF2u9BOLs8zMiM065X2uBa2xs2PS9DpFd68aSNe8xQKR11t9bIyi19/Klm+yWIYMbW6W6lYl3dPJZhoswAtyPMlo8TyufAyUaFZTT6OkJpzSKUz+fa+P5XE4pnWWAn27+v20fpp63zS7EQ9k5GyjrXns2Xfjq6UXZKl6wSLyDWO55NrHfOBB5yzcDjTlDELO1NxAxdTSjyi10jQ4SkPxB1gNpW/BpJx3MJ4eByH5HWp1K27F4ybGs6RxJrEF8nRpJvI7zGKlMW9Q4ggjPvIUZjvyOURx2OBm4kAtn2JwLI6rjbduNY1gD2ItJiNE/bK+fImk5I2pD4Hf6HiGlolEg+m8gFNi7F5PHNMuv5KBHcWEnpjOGIsByxkP5eSbUcEjo5v4oJ6gZHxP0hk31+gn7GvFESpuCJ2nUtNsX2fGLg0Faga+FbwOsJU+Vswkmm1+le9iayKlxie6uy7vj5mn38kWpBzhNxiqTliUq/VxL/02Ed9jklDJvSWdL2WcXr63Zgv7IW8OkafXyWuAjPdCSW3IqoBV+XtW3vImlH29+XdByjK19ORjeO5efEOsjJVPc2mwqvr44px34/cTz7eCSp46WKvGETtX02McPZiRgB/gioY8ZpdUSok/q+lbbPL0e9lJuanQDapN0bZccOGZJ+bPt1ZU1nrHOrVsCu7ZtVaskAR0q6CqhTHqFTHrb9sCSKJ92NkioHyRY+DlykqHPUHPBb5zr7crkmfqqoiroENbJM9JJhUzrdqDXyqO37JM2QNMP2OZK+0YYsDc+q+yWtT0Ty13FL7caxPGz7mzXbQBe9vhQ1Zz7GwmbCOgkqAXaxPab3n+3XjrW/iSMIRfNKt1Fm2nY3PaI6Pb9WIDwCL2MkCNC2d6zRR1s3yi79DnuX5+260Ne/FMHOV0v6MlHFs5Y1oAvcWWYYPwfOkPQPoG7g7XeJ2fp11BgctvAzSa+2/ajtR4r59JdEPNu0YtiUzt6EB8f7iBHl1sBbavZxf1mEu4CIcbmHpgjfGhxa/vj/R6Q9Waa8rko3juXA4kzwa0aPoCb05LJ9cnn+PizwqrGrR0c3cyxxw38VETf0FqJGUF3+rnBBX53RpsJJR4PuUp34MqPaj5EMz43ca3XW2Rrn1/m0d341n0MNz7PX12gPHd4oOxlI2P5reb699DXKY6smbyKUzHuBDxApm3Zqs6+2sP2a8nK/YrpdjvphGk9w58lnfw78WNLOxO9wElEnadoxVI4EDTq5SSpSzjeyIe9GnETHukbW2G7S4bF8gbgw/8jICMpVZxmSZhNrS8sSv8f9wP+6XkbjRnnlBbnOVMoMVz8SkNQoTjVqNDjR6HsCU047JRqQdCNxc2vN8Fwno/DSwL+Jm2Vb55cimPQNRMqVW4myEe3mHHxJkeFXbkqZNEmbXxMDiQ/TNJBwvbpC7yRKiLftsVX6WRx4VunjpqrH0E3KWtuTGT0Yqmz+lPR5IgvBydTMp9fSz3uAbYiB2TtdMV9krxkqpdNyk4RILFnrJln6aTtte1MfTyJGxS8s/VxAFFOqdHPpxrEoElSu1+6FqMjE/B7bF5TtFwEH17lZS7rE9gsknU7ELf2FyOe2Vk1ZaqXeL22eavuv6kKJhtLfpbar5Jur2t+KwH2ucBEq6ujMKY9GoswP2x7z2Cr0twnhhGEiCWzlOKZuDCS65LH1KsKh4o/EQGIN4mZ7Wrt9tiHDXoTDz92MHtjVuUZuHWN3JQWs0bF4IpwpriXSFPU0Jq8qw2ZeOwJ4d8tN8kii2mMl1GHa9iaOJ0wojen+bsSN4mXjthhNx8dCJNZcnvCSaofHG98PYPs3kipnhi58TpF5+0OEe+ssYrZQl6MVEeunUHE02GrK6QLnKOJLfkYNcyUscO/9IpEu5bOEd+OKRIaCN9uezCRzIzFw2c6lCqSkdn5HJH2amCU10vccKekEV3chb6xX/rXc+P8C/NcEnx+LbnhsHQBs1fR7rEWsY/RM6RBm8HU7sYTYrl0ksonW1EM/G2f/tGHYZjpXuaU0ct0RsrqQtr20u94tkeqSrrO9QcX23TiWcwkldTk1ygprpBzBm4maNT8kRsS7EovQ/Sh+9h7g/wgTXyVzjKKA3EQuvnULyZ3TaNrYRUVzpaLMxCcIU9ahRGaDS4rn2A9b/+sx2r+aWLt5IbFmcDwR19JOVdubgI3cVGeJSL5atYz5doQCfAYjA4nP2K5cG6eYCI8E2vbYap1dSRIRhV/LdNsJ5Zx4uWuU6Rinn/VZeI3sBx2KNy0ZiplO003yPEWusuab5Lk1u+tG2naAX0t6PSPloXcm0utPSJePZcyMBBU4YIJ+Ko1S1H236w8RbuiVzTEuBeQkfZbwbDqakbW6p9b8fhj79686alvMpYxDmTlfUmS8URUSn9r+OfDzsia0I5EAdWVFXrsTXb1EBMTMZAliPQUiSeafqza2fUp5+QAjmRXq0rbHlkaK4M2VdCpxjZmYvV3epjztcgtwrqRfMlp5VjZrFWefLQmlcyqwLfAbIuNF1T7OYezrrK6X6JQzFEqH7twkG6P3m4nYj1+UtjtSo1550+haxI2hUfd8BvBPJvco6fhYihwzge+6jdK97k7G2m4n27yZ9s0xO9jeqGn7O4qyz3ViniD+vwZLEG6/VQN+m2+s/255r06yzYeA44DjinfkLoQn2aRKp2kA8ADhdn1G2X45cFmF9t0cSHTisbV90+u7gZeU1/dSrwprN/hTeSxO/SzwDXYGNiJKNbxN0pMZuW9Upfm+sgRh1u9o9jVVDIXS6dJNsmEDHS9te1VZOrKldulYuhJIWNZi9mWkdvt5wP6OQnWTff/JRfFtYLsbrpsPEfEY51DfHPOQpN0Ik5SJxfjabvC2Rw0IFHnQqiaY3EjSg8RgZMnymrLd1o3Skfj00PKoQmMAcAVRebTBuRXbd3MgcZoi51htjy3bb+vC93cFlzyGaj+TO8C/HaVPHiveqvcQpss6crQ6GF2oiOWadgyF0mnQ4U1yVBLMDk+iRh87NMlybpNZokrbto+libECCeukwTmCcEZolDF4E2GHnywYs/E9j2uSbAE1+Hl5tMMbiAJsB5bt35R9nbIUkUpmUmz3PTrcTe7lxdW4kTKmSpmIbg8k5pTn5uwBpkblUElPJ9aUGufYBcDetu/sULbKlLWYoymOFJL+BrzZTTWkKjBXETf1PWJA8E8iLVEdOZodOWYQQaHL1emjVwybI8FPiZtk4+J6E7FgWukmWfoYdRIR7ql1TyIUNeA3JYIjIS6yubYrpejo0rG8ZKz9jrLFVdpfbfu5k+2bpI/vAKsQ+bGaFV+dwmeNvmrfKLuJRsf7zARWIgYC3+qlHJ0iaUvivLqNmGk9A3iLKyb8lHSxuxRw2wnFPHgccb1CpBPazfbLeyjDRcAnbZ9TtrcEPm97izb7Wx2YZbuySb+0u5URs/5jRPzW/rZ/044cU8mwKZ1u3CS7chIpYlye60hs2FhjucoV/fe7cSydIuli4CONE7fMWr5a54ajkQJbzdgVs2039bMlbd4oJa1JzHJeQFyYFwMf8DhpdSbopzkm5jHg7k69lvqBIo/dG2zfVLafSXjQVUqZ0o2BhKQnAO+iyRJArEFWHkhMk2vkmpb1wjH3jdN2Qk9Ud6EG1HRkqMxrwL8lvajlJtm6aDsZSzcUDoDtc4vHUDssz0gZ27pT3baPRdJvbL9oDJfhhotvVVfhPYEfFFMfwD+omYqni/b3A4BXtN4oqZZb6jjg20AjZcnrS9tagZ7uXrxPv3lC43cEsP37ogSqsgTh0dnsGWXqle3+DvAERiqFvqnse0eNPu6T9Ebiv4SwJvQ6c8gtkv4fo2dbVQczrU5DzZhJsn7DiHNHeb2L7ROa3vu87U9UlKV32B6aB+EBcg0xGr6diMrdqGYfJxL5rVYvj08RLql1ZZlTZDiKGKHfCuxao/1zm47ltjrHAqzWhd9yJjGrgYjDmNVmP88kqqZeX7Y3BD7VRj/XVtlXo+01U3UeTvcHsVZ3GOGmuyWxlnBEj2VY6Pev+58QOfBOIrzW7iHW/Fbt8XGsQGTauLI8DgRW6OH3XznW67G2p8tjqMxrDYoHCC5lhWu2XYHICdVIEXIBEfj2jzb6eiqxrgMRtNZOOp3ax6KmIFJJP7XdVhJElRQ27bRt6uM84COE6WTjsm+hwNkK/RxBuB03XEl3A2Z6AjNd0+Lqx4hZWsN7bVfixtDLFPjTBklPBN7DSAnxC4j0Ro+M32pU+2cSs5In215fUedoB9coiifpSiJz+B/L9ppEeqRaqY6GBUlbsHAy20njdNQURK6WgPLW7enCUJnXysW0E+XPawTd2d6/ah9FubxP0tKOmIh2ZRFRFXFN2/tLWlXSZrYruTEWX/3PA0+zva2k9YhcVYdXad70upNKo1dJOonOnACWsn2ZRgdAtrMO8i7iRtlwkb6AMJlNxBWMLK5CVKpsYHpbd2Xa4Eh9/x3gl24ys9Xge5SBROnvWkW9qDqVWD9CpBW6hfh/ViMK9E3KeHFCDdzDek+KHImfYGGFUSf11tHAWkRBvUYiWVMtONTjvB5re1owVEqHiKl5gLjZVBq1tVJGHIcRpQhWlbQRkUTw3TW7OpgYmW9NVA6dR1S7rJqi4yjCPfmTZfv3RO62KkpnohOxDt2w3f9NkRPLAIrU639tQ5Y9HVHeCyK9Je3NiBv0QriznFZDS3Hl/woRzLiGpOcSnk5VXek7Hkg4ChSuAzRS79xUdabFSJzQC4ko/h+V7V2A39aRowscSyjQTmrhzCYS87ZzrXY9/mvK6bd9r5sPyrpBh31cSnhGXdVJvxR7aks/lW3WRHbr1vZXV2z7OPAgoegeK68b2w/2+D9ZEziTyCbwZyJGZrV2f8+WfVdVbPvmsR69/B2m04MYlC3Xcm5dV6P9acTIvHGO7wycVlOG9wDLN22vQCS4rdPHJUR6ocb2E4hqqL38LX/ThT5OAJ7a7/OiV49hm+lcJGkD29d10ontO1pGcY+P99kJeLS4STdG+CtRbyT0kCLZaKP9C4hZ3KS4S4GIXQq+s+2XFQ/AGbbnSao8A5E0hwjkXKOY+hosy4hn4GQ0zy6XIMyeV1Ijt9WQ8ajtB1rO8Tqj7PcQWRCeJenPhJPMbjVl2N32AvOo7X8osogfPEGbVlYgnFwa58EyZV8v2VfSYYSzTHNmhUmtAZJOJn73ZYHfKoK4KyfmHVSGTem8CHhrCZR6BNoq1nVHMbG5uJHuTfX8Ws18k/CEW1nS/xGjwTqVQz9IeOasJelCIhBxlzbk6IQjCXfjxve+seyrE3z3U2ATj14f+wnVy+heRJjjVmS0i+k8KubEs71X87Yi+vv4it8/jNwg6Q3AzGLieh/xO1elo4FEYaYkuQz1ywCtbu6yLxLrjucQ1/p/EzWsesnbiCJyT6Cpng7VTNAnEcXfLmjZ/2LaM0EPBEPlvaYuFOtSFNY6kKh7IyKR4vtcs4pf6etZxKhawFm2Kyuv4hTxOGHzFnATcYG3tVbVDp0E35Vjfw7wZcLm3WAWEXD6nJqyrAn8xaPT8T/Z9m11+iltn0CYTGuVqxgWJC1FrBW+ouw6nSgwWNV7baESGyqF3WrI8BXCeeC7Zdc7gTtsf6hqH6WfpzASb3Wp2/AQ7QRJN7V7Hkk6Bfh4q2VG0gZEQPr2Y7ccbIZipiNplsOluHZJ5zFY1/YoU0EJzLywpkxH234TUXyrdV8VLi4X9oL0O8XNtJcupZ0E361LZGFentFZgecBu7chy4+B5qwQjxO28EkdM5rMGBB5qdZjpOTEosirbH+SEScVJO1C/J7j0jSQWE4j5QUgBhJ1F60/Riiad5XtMwgHnkmR9CxHSYjGtXBHeX6apKe5t5H8F0laz3Y7DgxPHmspwPZ1inQ4Q8lQzHQknWJ7O43OP9TArlF3fZxRXDulkke1KeaD62yvN0m7pxApRo4h1jIaxzILOMRtlCpolzJzPAjYnPhdLyJmfXXqv29uu1bywnH6GWvWVTXdSHMOuseA22uuSw0V7Z7jknYEXg3sQJiGGswDjrddx0TXNpIOtb2HRorqNWP3sIaMpN8RThW1TfqS/mB7nXHeu9n22l0VdpowFDMd29uV57ZdZCVtToykV9LouuOziOj8qv18nPDbb7gvNpTGf6iWgv6VwFuJ7MXNhaDmlX57RjFLdrqYuaek39m+HxYE3x7gmrnXgHsl7eBSnbLcAKsWdJvLSPr4ZwKbSLrbPU4Y2m8kbQv8D7CKpG82vTWLCi7Ptn8B/KIbA4mmAWLrd0w6QLS9R3nuShmQDtmmg7ZzJe1u+3vNOyW9g/AwHEqGYqbTQNJZtl862b5x2r6ESAmyJ3BI01vzgJNt/6GmLF9wBxHvknay/dN223dCN4PvNHbZ7YX2VehnLSIm4mmEIr+DcHu+uULbK4jF2RUIM+nlwH9azajDTok5ey4RN9ZcwG4ecI4rZt2Q9H3Ci/H+sl17IFE8MxssQTir/JftyoX1FCXMj22RY47tOh5wbdEw6Wt0SYEFVFkDVgSAn0gMSBtKZjbhUPGaXq9P9YqhUDqSliBqm5xDKI5mk9Sv6pikJK1m+3b1v57O8sSNoZN6Om0hacKknm6qy1Khr2uALRs3tHKRnmd7gzZlq/2/NExHkvYClnRUv6zkEDGMFEcK0WaZiG4NJMbot64zwlgm147lqPjd3TTpbwU00kLdYPvsLoo67RgK8xqxIPl+YhR8BSMnwINA3Vony0q6itFFmd5i+/o6nUj6ArAZI/V09pa0hatnfT2cDgqodUIdpVKBA4CLJZ1A/C87A/9XtxN1luJIxXy6G/D2sq/vRdX6yBZEjNJtxH/yDElvccV6OsAMSSu0DCRq3Us0Oq3/DGKEX/d+1A2367bohkm/qa9ziAHzIsFQzHQaSNrL9kEd9pH1dEa+75lE7fXVGZ1XqtZCraTnAA37+9ntePpI+hUjKY4WBOu6pYT0OG3/mziOC21/SeF+/f46ZsJhQp3X03kzsb44aiBh++gJG47uo/km2yg6doBr5ILT2G7Xf3J3yqNXlaFtk/6iyrDMdBrMl7R8hzbega+n00VOINa3DqO9rAwA2L5B0r0Ut1pJq9bxgCs83XZbi7ZlBH9+0/YtjCQOXRTpqJ6O7R8UxdUYSLy2zkBC0gzCE/NHk354Yj4G7EGsw0IECz+lwz4r0WTSX7HcZ5pN+qv0QoZBZdiUTjdSa3RSlKmZz7NwtPQ+Ndp3XECtCzxm+zuddFDWtQ4gTJ/3ECPT3xHxHnXoSoqjBIArFKlbmstEzJ3g8wvRyUCieBF+hJFEnW1R+rmUcFl+HZG1olfON9006S9SDJt57TpgwxYb77WuEf2uLtTTKSO5nUvb2vV0itxfsv1hdVAbqF2aPHLeRyiKExmdE6pydobiSLA1cKbtjcui6Rttv32Spq39/BZYmzbiIZLRqPN6OmMOJGpeZ18kXN5/xOiyGVW8vp5JBCrPaerjw7bHzEgylXTDpL+oMWxKpyMbb7nZn+ku+P9Lmmt7dgftOy6g1sF3j+WR06CuZ85c27OL8tm4jE4rBXW29NNxiqNkwTl+Qx2PzjH66HggUc6xViqdW5LmE4ry7S4u85JuqXNedhNJ6xNZLhZkZXCFAmyLKsNmXuvIxmv7cUnzJS3XBdfkMyV9mDZGcoVuFFBri2545DRxf3FzvgA4VtI9NB1PDZluB5C0MjVTrigyfO/Owg4RdQNUB55yjt/U5rpag0dt3ydphqQZts+R9I2qjYslYJ8O1nReC7yeKAL3KyJ561gDpClH0r5EmMZ6wKnAtkT5jlQ64zBUMx0ASRsT6WNeR6zF/NR2ZRurpF8AGxO5oJpv9rUWnjsZyZX2R47TfspvlIp8a2r1RpL0JuBx28fV6Gtp4GHiprAb4VBxrO2qOdwa/bRt0ikeiRewsOdbX4Jv+42k84lz/DJGn+OVsk9IOpNIh/NF4EnE/7FpHQ/PTi0BpY+lgR0JM9vWxI3+RNu/7qTfmjJcB2xEeKZupAj4PMZ2nUzsixRDoXS6aePVOIGRXY5dmdaUxdmXuiUAs1zk51d1rW1q9xQiZslEcbrakdadmHR67Wo+3dHoXHQLsH1exfYdDyQ6WdMZp78ViKwGu/bSXVnSZbY3a/Lmm0cMhnqWI3HQGBal01Ubr6TFiRoZJqK1/9NGH0sA72a0Q8IhLqn5K7Rfkyix8ILS/mIitmSsGVRX0QTJHyVdW2fxXpFH6tPA2cRN6iVEZoUjasrU9tqQpM8BF9k+tc53DhvlnNyTcMi4Djjcdq0y0019dTSQ6NQSMF2QdDARs/R64EPAP4kKv2/rq2DTmGFROq8m/vQXAg0b72HtrE1I+h/CEeGPxE1yDeCdtk+r2c+PiVFPwy31DUR53kqF2CRdAnybkbICrwf2sv388Vt1B0Xm3NkeXXgNScsSN5g6aYVuArZojIIVObcucs0aJE0mnS8QrrGVTTqS5gFLEzmuGulebHtWHRkGHUk/Io7/AmLt4Xbbe7fRT1cGEsOGohzBLNuVigsuqgyF0mnQDRuvpBuB7ZpmTGsBv6w7XZb0W7eUMRhr3wTtF5pRtOP11Q7FAeKlwJ5NC/irE0rwXNtfqdHXRUTutf+U7cVLH3UzPHRlbWhRRtJ1LjnvJC1GuPHXrs/UjYGEopDcB4FVHWUK1iFqWVXOTzhdkLQhCzupTLnDz6AyVN5rZWR+HHBck433Y0T1z6rM8+jMxbfQXnG4KyW9wPYlAJKeT70AvNMk7UPM2gzsCpzaiKFp1/ZdBdtflfRP4PzieQZhNviiKwaLaqQ8xM3ApcVBw8SgoPZIsGXWVXt9TR0kXx0iFiT1tP2Y1LbD132MvibmUb24X4MjCceOxuDjz4Sn5kD9L5KOADYkii3WLVe9SDJUM51uIOk7hGfUj4mTZxfgT8CZUH0EU0xU65a2AKsSJacfo0JQ4zg27wZTavtWJMe8xLaLSQ3btRRvcSUdF9ufqdjPPEbHDDVO2EZw6KQmsrJovSkjyVfnAHPdQemJQUTS44ws2gtYEvgXFX/LpoHEc4ENgFEDCdtvrSFLY43uKpes0L2ayXeTOtaLJBiqmU6XWAK4m7BTA9xLXJzbU28E00lxp27HytTlzcC3Jf2eWCP7FTVne1WVSoV+lu1CN//D6OSr3weuAhYppWO708zajf/ij+XR4Bdt9PUfSUtSBhHFjF0pI8I042K1X656kSRnOlNMazCj65V67muks6RnEQvOryTWUM4hFNCFtislAC2BmR8lcq01H0elTNUtHlfXAkfU9bhSZPzesmGSLCbKc+t44SXdQVLDOWYp4JPE+f1rwgnorbbP7Z909Snu5ycBd5HpmSqRSqeFcpN7OwvfJGsFZXYSzFjajxnpbHvnOnJ0izIq3arIsXnVwD5Jv6bETRHK4y3AvbY/VrF9xx5XkuYQgYyjkq+68yzHSU0k7U14Yj6VCMD+E3AlcKntquXHpw2SbiYcIq5jZE0n0zNNQCqdFhTFxm4kXJz3JzylftfGja6j/FTTIdK5mDzutP2Ioq7QhsAPXEpHVOzjCtvPa/bGk3S57U0na1s+2y2Pq6fSRvLVZGpQ5NJ7fXksSTgAHW/7930VrCaSLra9eb/lGCRm9FuAacjatv8f8JAjC8GrgHZiYx4tLqUL8lMR1RGr8u+yBvGYItP0PcAz2pCjE34KPC5pbeDQ8v2VU+AUGh5Tf5X0KkWaojHryk/SnjbMaps0HsTI+s7yeJpGV65MKiJppqQPdNqP7dttf6k4EcwBXkOUvBg0rpJ0nKQ5kl7bePRbqOlMOhIsTOMmd39ZU7kLWLmNfhqJLs+nvUSXcyUtD3yPcC39J5GVoJfML661rwEOsn2QopR3HT6nqAn0IeAgoshVnZvWRpIaZR0ELFm2q3hcTVRV1MRMNKmBI2HoHODrnfRTZq3bEjOdlwLnAvt1Kl8fWJJYy3lF0750mZ6ANK+1UKKtf0qYko4ElgE+bfuQmv0sTVT6nMFIMOMx7cTX9CvSWZGD7RvEgu/2tm+VdL3t9Su2nwm8z3ZHN6hkeiHp68ATWDhv2pUV2r6cmNn8D5Fw9HjgF63ZL5LhJZXOFCHpS62L5WPtm6SPvkY6S1qPWPy/2PYPJa0BvM72l2r0cZntzaZMyIm/e2vbZ49n7sio8fZQVMNtxVU8EiWdTZhof+oahRGnK+Wa2IuFr9NKGbsXRVLptKCoqrgTC59E+9fsZ6GkmWOltpmg/ZiRznW96DqleK2tavumNtu3PSruFEmfsb2v+lgmIhluisPQ4SzsvVYpY/eiSCqdFhRFoR5g4dorE60PNLd/F5Fdek1GB9AtS8S3vLFiP32PdJa0PfBVYHHba0h6LpHYsfIorpNRcTI9KZ6UnweeZnvbMiPe3PbhfRat50i61D1IwjtMpNJpoc6axTjtlwNWILIh79P01rw66zmSDgcO6Geks6JGyNZEIGUjVcnAremU2JAjiawK3wM2IeJ0elbsa5iQdBrxe36yuPMvRrj2b9Bn0XqOpDcA6xABrgsyKvRiJj+opMv0wlwkqe2Lx/YDtm+zPYdwMd66BIrNKPbfqvyASLFxk6RrJV1XIut7yaNeuGz3/DE/OQYla8Gc7orUFv9r+0HCw+hJwJuIYNGkPVa0/WPKuVBc2StlqBhCNiBKoX+R8JY8gLAOJOOQLtOFEoxp4jd5m6Rb6CCtRckoMJtI+nkksDhRW+eFFbs4nLg5jrIV95gbykhupiL1/PuAi2r2caGkb9GHNZ0mGslC/4cIbr1Baj/FcsJDinIGjbxpLyBM0osiuwBruo1Cj4sqqXRG2K7L/b2GqEN/JYDtv6hkbK7IvbZP6rJMddmLcJd+hPA4Oh34bM0+nluemx0xeh0jc0VJx7MG8PHyP/RLkQ8DHyTyja0l6UJgJaAv6ZmmAdcDyxPB20kFck1nDEq0eqPM9IXtjMo1Ujv9StublLidi2t4rx1MnMwnM9pW3EuX6V1snzDZvumOpBmE8rvF9v1llL5Kr+OehomyjrMuMYu8yfajkzQZSiSdS3iZXs7o6zRdpschlU4Lkj5NTJkbN/dXAyfY/lzNfj5MLDC+nHAq+F/gONsHVWzfdzffcdy+F9o3SR/TwtOpxOo0BhK/sX1iL79/2JC0BQuHFfQsA/p0QZFleiHSZXp8Uum0oCjFu5Hth8v2ksDVrlGKt6mvlxOL1wJOt31GV4WdIiRtS6x/vI5Yi2kwC1ivTrDndPB0KrPGtYmU+hBVWP9o+z29kmGYkHQ0sBZwNSMOBLb9vr4J1UdK8tJ1bJ+pKMM90zWLHi5K5JrOwvyFKGnwcNl+IlFKtzZFyZwhaUVqlvPtc6TzX4jS2jsQ8UoN5lExb5qkxYpX04q2fyzp47CgTHKvPZ22Bp7tMsJSFHHLolvtM5sYfCzyI1ZJuwN7EEls1wJWAQ4h8sklY5BKZ2EeILy2ziBMMS8HLpP0TYDJRnPFk+eLwN+JRfejgRUJl+k32/5VRTl+TniwnUyPF71tXwNcI+k44hxpJyPBZUQ8zHTwdLqZKBfeqHHyDOAPPZZhmLgeeArw134LMg14D7AZcCmA7T8oCjcm45BKZ2FOLI8G59Zs/y3gE0SCz7OBbW1foqjC+UOi8mYVHrb9zZrf3W22oWQkAOpmJGi4JE8HT6dlgd9JuoxQfpsRWbxPglz0rYqkk4nfb1ngt+X3XNQXzx+x/Z+GB34xHy/yM8CJyDWdLiPpatvPLa9/Z/vZTe9d1Yjsr9BP3yOdx8lIsKCo2iRt7wS+VjZnEGZKEcfyuO2vjde224y32NsgF32rkb/jwkj6MnA/8GbCHP5u4Le2P9lPuaYzOdMpNAWHjoVtb1Sxq2ZT2L9b+6kh0gZEcOjWTX32Or7lUdsPtMRRVj2GmURZiNYgzKW6IVgdbJ/Xsti7JLBYLvbWo6FUNE4GdWCRUzpEqqu3E0HcewC/tH1Yf0Wa3uRMp1BuSgvtJuz/H7f9PxX7eZyIvBdR4OlfTX0tYfsJFfu5mVis7Vukc8n/dhZxYe1EZCR4gu09K7St5Vo9lTQv9tpeq2RXOMR2Lva2wTiu9JUzqA8DknYEnm7722X7MsJ0bOCjtn/ST/mmMznTKZT8aAAoSiq/gYjXuZUo6la1n5ldEmk6RDo3ZyT4IfUyEkynNDO52NsFmjOot+QBXBa4sD9S9Y2PElVPGywOPI+Y3R8JpNIZh1Q6BUnPJJJTzgH+RsSnyPZWfRJpeeBGSX2LdLb9L0LptGOfnk6ziFzs7Q7HAafRYQb1IWFx23c0bf+m/AZ/L9lHknFIpTPCjcAFwHa2bwaQVCkmZYrYt19f3OSlNCZVFN80uwmdJ+kTwJIlYPfdhCt6UoOScfwBYI6kjYAXl7cuIEIEFiVWaN6w/d6mzZV6LMtAkUpnhNcS0+VzFIXcjqePJqI+ewI1UrO/lojHOKZszwHu7otEndG82PtO4FQgF3vbRNL7iDWyRqqoYyQdWjXF05BwqaTdbX+veaekdxIxask4pCNBC2VqvCNxg92aqGtzontc8KsEUR4EPJuwF88EHrI9q4cyzLU9e7J90xVJhxLmoDPTU617lPWczW0/VLZrJbMdBsqa4M8J03cjjOF5RGjAq20P4uCsJ2QRtxZsP2T7ONvbA08HrgI+NkmzqeBbhOL7A+EF9w7g2z2WYWlJazY2SmqeQbJXHw5sBJwq6SxJHytmoaQzxOiibY8zvRxHphzb99jegnCsua089re9eSqcicmZzjSlMaNodkWtE1zaJRm2AQ4FbiFuKqsB77R9eq9k6BYlFc8rgG2JVPRXAr9yVMBMaiDpg8BbGMnc8WrgKNvf6JdMyeCQSmeaIul84GXEaP2v5fHWGkGq3ZLjicCzyuaNth+Z6PODgqTnAdvY/r9+yzKINNWcArjA9lX9lCcZHFLpTFNKsOrdxHrOB4iyAt9peNZN8XdvbfvsUoNmIdzDQnLdQtKrgOcQGcQBsL3/+C2SJJkK0nttmjFGpPN5wMqEC/PFRMbkqeYlRLLS7cd4z4x4LQ0Ekg4h0u9sRXit7Ux6GCVJX8iZzjSjZGJ+fSPwTNLVhBfdMsCRmbqlPo11sabnZYDTbL940sZJknSV9F6bfowZ6Wz7T/TYc0zScpK+JmlueRwgableytAlGolX/yXpacCjwFP7KE+SLLKk0pl+TKdI5yOIaqGvK48HibxSg8YpkpYHvkJ4rd3GSOnqJEl6SJrXphmSjiXq14wV6byl7Tk9lGVBbaCJ9k13JD2x4XVXvPGWIIrkDYUnXpIMEulIMP34APDzUsRtoUjnHsvyb0kvsv0bAEkvZOEaQYPAxUTpbIqieUTSlY19SZL0jlQ60wzb9wBbSNqacPGFKAx1dh/E2RP4QdM6zj+IoMCBQNJTgFWIRJ8bMxI1P4s+FJNLkiSVzrSlKJl+KBoAJM0E3mR7I0mzikwP9kueNnkl8FYinVFzeex5wCf6IVCSLOrkmk4yLpIusf2CfsvRKZJ2sl25EF+SJFNHKp1kXCR9hzBPnUCU4AYGJyNByRE2Lra/NtH7SZJ0nzSvJROxBHAfEZzaYJAyEizbbwGSJBlNznSSJEmSnpHBocm4SHpmqUNzfdneUNKn+i1XXYblOJJkGEilk0zE94CPE2ljsH0tUdJ70BiW40iSgSeVTjIRS9luzcb8WF8k6YxhOY4kGXhS6SQT8TdJaxHOA0jamSgmN2gMy3EkycCTjgTJuEhakyhXvQWRjeBWYDfbt/dVsJqMcxxvtH1bP+VKkkWRVDrJuEhaw/atkpYGZtie19jXb9naofk4+i1LkiyqpNJJxkXSlbY3adl3he3n9UumOmRwaJJMPzI4NFkISc8iko0uJ+m1TW/NIgJGB4VGcOi6wKbASWV7e7JcdZL0hVQ6yVisC2wHLE/coBvMA3bvh0DtYPszAJLOBzZpmNUk7Qf8so+iJckiS5rXknGRtLnti/stR6dIugnYsKWQ27W21+2vZEmy6JEu08lE7FnKPAMgaQVJR/RRnnb5AXCZpP3KLOdS4Ki+SpQkiyg500nGRdJVtjeebN8gIGkT4MVl83zbV/VTniRZVMk1nWQiZkhawfY/ACT9FwN6zti+kpHy30mS9ImBvIEkPeMA4GJJJxClnncG/q+/IiVJMsikeS2ZEEnPAbYqm2fb/m0/5UmSZLBJpZNMiqSVaYrPsf2nPoqTJMkAk95rybhI2kHSH4hcZecBtwGn9VWoJEkGmlQ6yUR8FngB8HvbawAvBS7pr0hJkgwyqXSSiXjU9n2EF9sM2+cAs/stVJIkg0t6ryUTcb+kZYDzgWMl3QM81GeZkiQZYNKRIBmXUgrg38SMeDdgOeDYMvtJkiSpTSqdpBKSVgTuc54wSZJ0QK7pJAsh6QWSzpX0M0kbS7oeuB64W9I2/ZYvSZLBJWc6yUJImgt8gjCnHQpsa/uSUmfnh4OYey1JkulBznSSsVjM9q9tnwDcZfsSANs39lmuJEkGnFQ6yVjMb3r975b3cmqcJEnbpHktWQhJjxOu0QKWBP7VeAtYwvYT+iVbkiSDTSqdJEmSpGekeS1JkiTpGal0kiRJkp6RSidJkiTpGal0kiRJkp6RSieZtkjaT9KH22x7lKSduy1Ty3dcNJX9J8kwkkonSdrE9hb9liFJBo1UOklPkbS0pF9KukbS9ZJ2lXRbSSiKpNmSzm1qspGkiyX9QdLuE/QrSd+SdJOkM4GVm957qaSrJF0n6QhJTyz7b5P0BUlXS5oraRNJp0v6o6Q9y2eWkXSWpCtL+x2b+v1ned6y5Kr7iaQbJR0rSRPIuqmki8pvcJmkZSUtIenI8h1XSdqqfPatJQfer8pv8OWyf2aZzV1f2nyg7F+rfPYKSReU1EWNmd83y/fe0pgFSnqqpPPLb3C9pBc3H1t5vbOko8rrXcrnrpF0/uT/eJK0YDsf+ejZA9gJ+F7T9nJEGewVy/Zs4Nzyej/gGiJAdUXgDuBp4/T7WuAMYCbwNOB+YGdgidLumeVzPwDeX17fBryrvP46cC2wLLAScHfZvxgwq7xeEbiZkfi2f5bnLYEHgKcTA7mLgReNI+fiwC3ApmV7VvmODwFHlH3PAv5UZH9r+fxyZft24BnA84AzmvpdvjyfBaxTXj8fOLu8Pgo4oci3HnBz2f8h4JPl9Uxg2eZjK693Bo4qr68DVmn+znzko84jZzpJr7kOeLmkL0l6se0HJvn8L2z/2/bfgHOAzcb53H8TyUgft/0X4Oyyf13gVtu/L9vfL59tcFKTXJfanmf7XuARScsTWRg+L+la4ExgFeDJY3z/ZbbvtD0fuBpYfRw51wX+avtyANsP2n4MeBFwTNl3I6FcnlnanGX7AdsPA78FViMU0ZqSDiqZvx8sBfe2AE6QdDXwXeCpTd/9c9vzbf+26RguB94maT9gA9vzxpG7wYXAUWXWOXOSzybJQqTSSXpKuflvQtzkPyfp08BjjJyLS7Q2mWS7Ux4pz/ObXje2FyOK160EPM/2c4G7x5CxuR+Ax+luVd6F+rb9D2Aj4FxgT+Aw4je83/Zzmx7PHqcfAdg+n1DCfyaUyZvL+82/84Ljtb0n8ClitnWFpCd14fiSRYhUOklPkfQ04F+2jwG+Qiig2whzEYT5rZkdy3rHkwgz1uXjdH0+sGtZ63gqsFXZfxOwuqS1y/abgPNqiLwccI/tR8s6y2o12o7FTcBTJW0KUNZzFgMuIBQckp4JrFo+OyZlDWyG7Z8SSmAT2w8Ct0rapXxGkjaaSBhJqxGmxO8RimuT8tbdkp4taQbwmqbPr2X7UtufBu4llE+SVKabo7EkqcIGwFckzQceBd5FrNkcLumzxMi9mWsJs9qKwGeL6WwsTgS2JsxPfyLWVbD9sKS3ESanxQildUgNeY8FTpZ0HTAX6Ki8g+3/SNoVOEjSkkQW75cBBwPfKd/zGPBW249M4I+wCnBkUQoAHy/Pu5V+PgU8ATieWBcbjy2Bj0h6FPgn0Jjp7AOcQiiWucAyZf9XJK1DzJTOmqTvJFmITPiZJEmS9Iw0ryVJkiQ9I81ryUAhaQPg6Jbdj9h+fj/kmQhJJwJrtOz+mO3T+yFPkkwH0ryWJEmS9Iw0ryVJkiQ9I5VOkiRJ0jNS6SRJkiQ9I5VOkiRJ0jNS6SRJkiQ94/8Dy5IbyRVQ+DEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "data = consensus_taxonomy_votus\\\n",
    "    .groupby('sub_domain_consensus',as_index=False)\\\n",
    "    .agg({'votu':'count'})\\\n",
    "    .sort_values('votu',ascending=False)\n",
    "data['cat']=data['sub_domain_consensus'].map(lazy_coloring)\n",
    "sns.barplot(x='sub_domain_consensus',\n",
    "           y='votu',\n",
    "            # hue = 'sub_domain_consensus', \n",
    "            data = data,\n",
    "            # order = sub_domain_order,\n",
    "            dodge=False,\n",
    "           palette=lazy_coloring)\n",
    "plt.xticks(rotation=90)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
    "plt.yscale('log')\n",
    "plt.savefig(\n",
    "'/groups/banfield/projects/environmental/4thwedge/2021/4Wh2oV/figures_out/pHost/barplot_votuhost_phylum_update.pdf',\n",
    "    format='pdf'\n",
    ")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
